<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
    {
      "@context": "https://schema.org",
      "@id": "https://hds-sandbox.github.io/bulk_RNAseq_course/",
      "@type": "LearningResource",
      "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
      "description": "This workshop includes a tutorial on how to approach RNAseq data, starting from your sequencing reads (fastq files). Thus, the workshop only briefly touches upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, mainly for the purpose of outlining considerations in the downstream bioinformatic analysis. This workshop is based on the materials developed by members of the teaching team at the Harvard Chan Bioinformatics Core (HBC), a collection of modified tutorials from the DESeq2, R language vignettes and the nf-core rnaseq pipeline.",
      "keywords": [
        "bioinformatics, RNAseq, bulk, NGS"
      ],
      "learningResourceType": [
        "e-learning"
      ],
      "license": [
        {
          "@id": "https://creativecommons.org/licenses/by-sa/4.0/legalcode",
          "@type": "CreativeWork"
        }
      ],
      "mentions": [
        {
          "@id": "10.5281/zenodo.8046218",
          "@type": "Thing"
        },
        {
          "@id": "https://zenodo.org/record/7956067",
          "@type": "Thing"
        }
      ],
      "name": "Introduction to bulk RNAseq analysis workshop",
      "educationalLevel": "Beginner",
      "audience": "Researchers, PhDs, Postdocs, MSc"
    }
    </script><title>Ucloud setup to run a preprocessing pipeline - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.6543a935.min.css><link rel=stylesheet href=assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/neoteroi-mkdocs.css><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script id=__analytics>function __md_analytics(){function n(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],n("js",new Date),n("config",""),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&n("event","search",{search_term:this.value})}),document$.subscribe(function(){var a=document.forms.feedback;if(void 0!==a)for(var e of a.querySelectorAll("[type=submit]"))e.addEventListener("click",function(e){e.preventDefault();var t=document.location.pathname,e=this.getAttribute("data-md-value");n("event","feedback",{page:t,data:e}),a.firstElementChild.disabled=!0;e=a.querySelector(".md-feedback__note [data-md-value='"+e+"']");e&&(e.hidden=!1)}),a.hidden=!1}),location$.subscribe(function(e){n("config","",{page_path:e.pathname})})});var e=document.createElement("script");e.async=!0,e.src="https://www.googletagmanager.com/gtag/js?id=",document.getElementById("__analytics").insertAdjacentElement("afterEnd",e)}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#running-the-bulk-rnaseq-pipeline-in-ucloud class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Ucloud setup to run a preprocessing pipeline </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class=md-tabs__link> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class=md-tabs__item> <a href=05a_data_analysis_setup.html class=md-tabs__link> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=summary.html class=md-tabs__link> Workshop Summary </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> <span class=md-ellipsis> Course Introduction </span> </a> </li> <li class=md-nav__item> <a href=02_experimental_planning.html class=md-nav__link> <span class=md-ellipsis> Experimental planning </span> </a> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> <span class=md-ellipsis> Data Explanation </span> </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4 checked> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex> <span class=md-ellipsis> Read Preprocessing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=true> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> <span class=md-ellipsis> Preprocessing steps </span> </a> </li> <li class=md-nav__item> <a href=04b_pipelines.html class=md-nav__link> <span class=md-ellipsis> Nextflow and nf-core pipelines </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> UCloud setup </span> <span class="md-nav__icon md-icon"></span> </label> <a href=04c_preprocessing_setup.html class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> UCloud setup </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#submit-a-job-in-ucloud class=md-nav__link> <span class=md-ellipsis> Submit a job in Ucloud </span> </a> </li> <li class=md-nav__item> <a href=#understanding-the-pipeline-options class=md-nav__link> <span class=md-ellipsis> Understanding the pipeline options </span> </a> </li> <li class=md-nav__item> <a href=#restarting-a-failed-run class=md-nav__link> <span class=md-ellipsis> Restarting a failed run </span> </a> </li> <li class=md-nav__item> <a href=#stopping-the-app class=md-nav__link> <span class=md-ellipsis> Stopping the app </span> </a> </li> <li class=md-nav__item> <a href=#saved-results class=md-nav__link> <span class=md-ellipsis> Saved results </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> <span class=md-ellipsis> Analysing RNAseq data </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=05a_data_analysis_setup.html class=md-nav__link> <span class=md-ellipsis> Ucloud setup </span> </a> </li> <li class=md-nav__item> <a href=05b_count_matrix.html class=md-nav__link> <span class=md-ellipsis> Count matrix </span> </a> </li> <li class=md-nav__item> <a href=05c_count_normalization.html class=md-nav__link> <span class=md-ellipsis> Normalization </span> </a> </li> <li class=md-nav__item> <a href=06_exploratory_analysis.html class=md-nav__link> <span class=md-ellipsis> Exploratory Analysis </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_5> <label class=md-nav__link for=__nav_5_5 id=__nav_5_5_label tabindex=0> <span class=md-ellipsis> Differential Expression </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> <span class=md-ellipsis> Design formula </span> </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> <span class=md-ellipsis> DESeq2 </span> </a> </li> <li class=md-nav__item> <a href=07b_hypothesis_testing.html class=md-nav__link> <span class=md-ellipsis> Hypothesis testing </span> </a> </li> <li class=md-nav__item> <a href=07c_DEA_visualization.html class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_6> <label class=md-nav__link for=__nav_5_6 id=__nav_5_6_label tabindex=0> <span class=md-ellipsis> Functional Analysis </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_6_label aria-expanded=false> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=08a_FA_genomic_annotation.html class=md-nav__link> <span class=md-ellipsis> Gene annotation </span> </a> </li> <li class=md-nav__item> <a href=08b_FA_overrepresentation.html class=md-nav__link> <span class=md-ellipsis> Over representation analysis </span> </a> </li> <li class=md-nav__item> <a href=08c_FA_GSEA.html class=md-nav__link> <span class=md-ellipsis> Class scoring </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=09_summarized_workflow.html class=md-nav__link> <span class=md-ellipsis> Summarized workflow </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=summary.html class=md-nav__link> <span class=md-ellipsis> Workshop Summary </span> </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#submit-a-job-in-ucloud class=md-nav__link> <span class=md-ellipsis> Submit a job in Ucloud </span> </a> </li> <li class=md-nav__item> <a href=#understanding-the-pipeline-options class=md-nav__link> <span class=md-ellipsis> Understanding the pipeline options </span> </a> </li> <li class=md-nav__item> <a href=#restarting-a-failed-run class=md-nav__link> <span class=md-ellipsis> Restarting a failed run </span> </a> </li> <li class=md-nav__item> <a href=#stopping-the-app class=md-nav__link> <span class=md-ellipsis> Stopping the app </span> </a> </li> <li class=md-nav__item> <a href=#saved-results class=md-nav__link> <span class=md-ellipsis> Saved results </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <div><h1 id=running-the-bulk-rnaseq-pipeline-in-ucloud>Running the bulk RNAseq pipeline in uCloud<a class=headerlink href=#running-the-bulk-rnaseq-pipeline-in-ucloud title="Permanent link">Â¶</a></h1> <p><strong>Last updated:</strong> <em><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">May 21, 2024</span></em></p> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>ðŸ•° <strong>Time Estimation:</strong> X minutes </p> <p>ðŸ’¬ <strong>Learning Objectives:</strong> </p> <ol> <li>Learn about the UCloud computing system.</li> <li>Learn how to submit a job and explore your results folders.</li> <li>Submit a nf-core RNAseq run on our data</li> </ol> </div> <h2 id=submit-a-job-in-ucloud>Submit a job in Ucloud<a class=headerlink href=#submit-a-job-in-ucloud title="Permanent link">Â¶</a></h2> <p>Access <a href=https://cloud.sdu.dk>Ucloud</a> with your account and choose the project <code>Sandbox RNASeq Workshop</code> where you have been invited. Or ask to be invited to <a href=mailto:jose.romero@sund.ku.dk>jose.romero@sund.ku.dk</a>.</p> <p><img alt src=img/04c_preprocessing_setup/chooseProject.png></p> <p>Click on <code>Apps</code> on the left-side menu, and search for the application <code>nf-core rnaseq</code> and click on it.</p> <p><img alt src=img/04c_preprocessing_setup/chooseRNAseq.png></p> <p>You will be met with a series of possible parameters to choose. However, we have prepared the parameters already for you! Just click on <code>Import parameters</code>:</p> <p><img alt src=img/04c_preprocessing_setup/importParameters.png></p> <p>Then, <code>Import file from UCloud</code>:</p> <p><img alt src=img/04c_preprocessing_setup/importParameters2.png></p> <p>And select the <code>jobParameters_preprocessing.json</code> in:</p> <p><code>sandbox_bulkRNASeq</code> -&gt; <code>sequencing_data</code> -&gt; <code>Scripts</code> -&gt; <code>ucloud_preprocessing_setup</code> -&gt; <code>jobParameters_preprocessing.json</code></p> <div class="admonition warning"> <p class=admonition-title>Warning</p> <p><strong>Make sure that the hard-drive icon says <code>sequencing_data</code>!!</strong></p> <p>Otherwise, click on the down arrow (<strong>âˆ¨</strong>) icon and search for the folder.</p> <p><img alt src=img/04c_preprocessing_setup/importParameters3.png></p> </div> <p>You are ready to run the app! But first, take a look at the arguments used to run the job. We have given it a <code>Job name</code>, <code>Hours</code>, <code>Machine type</code> as well as an additional parameter <code>Interactive mode</code>. <code>Interactive mode</code> will allow us to follow the progress of the pipeline.</p> <p><img alt src=img/04c_preprocessing_setup/selectedParams.png></p> <p>In addition, we have also selected the <code>sequencing_data</code> folder to have access to our sequencing reads in our job.</p> <p><img alt src=img/04c_preprocessing_setup/selectedParams2.png></p> <p>Now click on the button on the right column of the screen (<code>submit</code>) to start the job.</p> <p align=center> <img src=./img/04c_preprocessing_setup/submit.png width=300> </p> <p>Now, wait some time until the screen looks like the figure below. It usually takes a few minutes for everything to be ready. You can always come back to this screen from the left menu <code>Runs</code> on UCloud, so that you can add extra time or stop the app if you will not use it.</p> <p><img alt src=img/04c_preprocessing_setup/startApp.png></p> <p>Now, click on <code>Open terminal</code> on the top right-hand side of the screen. You will start terminal session through your browser! Once inside the terminal, you will need to do one last thing before starting the pipeline:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a>tmux<span class=w> </span>
</code></pre></div> <p><img alt src=img/04c_preprocessing_setup/tmux.png></p> <p>The <code>tmux</code> command will start a virtual command line session that is <strong>recoverable</strong>. This is very important because once we start the pipeline, we will lose the ability to track the progress if your computer loses connection or is in sleeping mode. You will know you are inside the tmux virtual session by looking at the bottom of the screen:</p> <p><img alt src=img/04c_preprocessing_setup/tmux_session.png></p> <details class=tip> <summary>Reconnect to tmux session</summary> <p>If you want to leave the tmux session, you can do so by pressing simultaneously <code>Ctrl</code> and <code>B</code> keys, and then press <code>D</code>. Then you can reconnect to the session using the command:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a>tmux<span class=w> </span>attach<span class=w> </span>-t<span class=w> </span><span class=m>0</span>
</code></pre></div> </details> <p>We can finally start the run! Type in the command:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a>bash<span class=w> </span><span class=m>778339</span>/Scripts/preprocessing_salmon.sh
</code></pre></div> <p><img alt src=img/04c_preprocessing_setup/startRun.png></p> <p>This will run a small bash script that will start the nf-core pipeline. It is usually a good idea to save the command that was used to run the pipeline! You should see now a prompt like this, which means that the pipeline started successfully!</p> <p><img alt src=img/04c_preprocessing_setup/nf-core_start.png></p> <p>Inside the <code>preprocessing_salmon.sh</code> script you will find:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a>cp<span class=w> </span>/work/778339/raw_reads/samplesheet.csv<span class=w> </span>/work/samplesheet.csv
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a>cp<span class=w> </span>/work/778339/Scripts/ucloud_preprocessing_setup/nf-params_salmon.json<span class=w> </span>/work/nf-params_salmon.json
<a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a><span class=nb>cd</span><span class=w> </span>/work
<a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a>
<a id=__codelineno-3-5 name=__codelineno-3-5 href=#__codelineno-3-5></a>nextflow<span class=w> </span>run<span class=w> </span>nf-core/rnaseq<span class=w> </span>-r<span class=w> </span><span class=m>3</span>.11.2<span class=w> </span>-params-file<span class=w> </span>/work/nf-params_salmon.json<span class=w> </span>-profile<span class=w> </span>conda<span class=w> </span>--max_cpus<span class=w> </span><span class=m>8</span><span class=w> </span>--max_memory<span class=w> </span>40GB
<a id=__codelineno-3-6 name=__codelineno-3-6 href=#__codelineno-3-6></a>
<a id=__codelineno-3-7 name=__codelineno-3-7 href=#__codelineno-3-7></a><span class=c1># Search for the last file created by the pipeline, the multiqc_report, recursively</span>
<a id=__codelineno-3-8 name=__codelineno-3-8 href=#__codelineno-3-8></a><span class=nv>file_path</span><span class=o>=</span><span class=k>$(</span>find<span class=w> </span>/work/preprocessing_results_salmon<span class=w> </span>-name<span class=w> </span><span class=s2>"multiqc_report.html"</span><span class=w> </span><span class=m>2</span>&gt;/dev/null<span class=k>)</span>
<a id=__codelineno-3-9 name=__codelineno-3-9 href=#__codelineno-3-9></a>
<a id=__codelineno-3-10 name=__codelineno-3-10 href=#__codelineno-3-10></a><span class=c1># Check if the file exists</span>
<a id=__codelineno-3-11 name=__codelineno-3-11 href=#__codelineno-3-11></a><span class=k>if</span><span class=w> </span><span class=o>[[</span><span class=w> </span>-n<span class=w> </span><span class=s2>"</span><span class=nv>$file_path</span><span class=s2>"</span><span class=w> </span><span class=o>]]</span><span class=p>;</span><span class=w> </span><span class=k>then</span>
<a id=__codelineno-3-12 name=__codelineno-3-12 href=#__codelineno-3-12></a><span class=w>    </span><span class=c1># Clean run if the pipeline is completed</span>
<a id=__codelineno-3-13 name=__codelineno-3-13 href=#__codelineno-3-13></a><span class=w>    </span>rm<span class=w> </span>-r<span class=w> </span>/work/work
<a id=__codelineno-3-14 name=__codelineno-3-14 href=#__codelineno-3-14></a><span class=w>    </span>mv<span class=w> </span>/work/nf-params_salmon.json<span class=w> </span>/work/preprocessing_results_salmon/nf-params_salmon.json
<a id=__codelineno-3-15 name=__codelineno-3-15 href=#__codelineno-3-15></a>
<a id=__codelineno-3-16 name=__codelineno-3-16 href=#__codelineno-3-16></a><span class=k>fi</span>
</code></pre></div> <p>You see that we have copied the <code>samplesheet.csv</code> to the working directory <code>/work</code>. This is because the paths inside the <code>samplesheet.csv</code> for the fastq files of our samples are relative to the <code>/work</code> folder! <strong>It is very important that the fastq paths inside this file matches properly to the paths inside your job!</strong></p> <p>We have also copied the nf-params.json file with all the options used in the pipeline, so that you can find and replicate easily this run in the future. Finally, we remove the nextflow <code>work</code> directory if the pipeline completes successfully. this will save quite a bit if storage in the future, since the nextflow <code>work</code> directory will accumulate over runs.</p> <p>Then we are making sure that we are inside the correct folder before starting the job using <code>cd /work</code>. We will see in the section below the arguments we used to run the pipeline:</p> <h2 id=understanding-the-pipeline-options>Understanding the pipeline options<a class=headerlink href=#understanding-the-pipeline-options title="Permanent link">Â¶</a></h2> <p>Let's divide the command into different sections. First we have:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a>nextflow run nf-core/rnaseq -r 3.11.2
</code></pre></div> <p>While usually one would run an nf-core pipeline using <code>nextflow run nf-core/rnaseq</code> and fetch the pipeline remotely, UCloud has installed the pipelines locally. Specifically, we are using the version 3.11.2 by using the argument <code>-r 3.11.2</code>.</p> <hr> <p>Second, we have:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a>-params-file /work/778339/nf-params.json
</code></pre></div> <p>The <code>-params-file</code> argument is another nextflow core argument that allows us to give the nf-core rnaseq pipeline arguments in a <a href=https://www.json.org/json-en.html>json file</a>, instead of creating an excessively long command. Writing the parameters this way allows for better reproducibility, since you can reuse the file in the future. Inside this file, we find the following arguments:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=p>{</span>
<a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=w>    </span><span class=nt>"input"</span><span class=p>:</span><span class=w> </span><span class=s2>"/work/samplesheet.csv"</span><span class=p>,</span>
<a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a><span class=w>    </span><span class=nt>"outdir"</span><span class=p>:</span><span class=w> </span><span class=s2>"/work/preprocessing_results_salmon"</span><span class=p>,</span>
<a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a><span class=w>    </span><span class=nt>"fasta"</span><span class=p>:</span><span class=w> </span><span class=s2>"/work/778339/genomic_resources/homo_sapiens/GRCh38/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz"</span><span class=p>,</span>
<a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a><span class=w>    </span><span class=nt>"gtf"</span><span class=p>:</span><span class=w> </span><span class=s2>"/work/778339/genomic_resources/homo_sapiens/GRCh38/Homo_sapiens.GRCh38.109.MODIFIED.gtf.gz"</span><span class=p>,</span>
<a id=__codelineno-6-6 name=__codelineno-6-6 href=#__codelineno-6-6></a><span class=w>    </span><span class=nt>"pseudo_aligner"</span><span class=p>:</span><span class=w> </span><span class=s2>"salmon"</span><span class=p>,</span>
<a id=__codelineno-6-7 name=__codelineno-6-7 href=#__codelineno-6-7></a><span class=w>    </span><span class=nt>"skip_stringtie"</span><span class=p>:</span><span class=w> </span><span class=kc>true</span><span class=p>,</span>
<a id=__codelineno-6-8 name=__codelineno-6-8 href=#__codelineno-6-8></a><span class=w>    </span><span class=nt>"skip_rseqc"</span><span class=p>:</span><span class=w> </span><span class=kc>true</span><span class=p>,</span>
<a id=__codelineno-6-9 name=__codelineno-6-9 href=#__codelineno-6-9></a><span class=w>    </span><span class=nt>"skip_preseq"</span><span class=p>:</span><span class=w> </span><span class=kc>true</span><span class=p>,</span>
<a id=__codelineno-6-10 name=__codelineno-6-10 href=#__codelineno-6-10></a><span class=w>    </span><span class=nt>"skip_qualimap"</span><span class=p>:</span><span class=w> </span><span class=kc>true</span><span class=p>,</span>
<a id=__codelineno-6-11 name=__codelineno-6-11 href=#__codelineno-6-11></a><span class=w>    </span><span class=nt>"skip_biotype_qc"</span><span class=p>:</span><span class=w> </span><span class=kc>true</span><span class=p>,</span>
<a id=__codelineno-6-12 name=__codelineno-6-12 href=#__codelineno-6-12></a><span class=w>    </span><span class=nt>"skip_bigwig"</span><span class=p>:</span><span class=w> </span><span class=kc>true</span><span class=p>,</span>
<a id=__codelineno-6-13 name=__codelineno-6-13 href=#__codelineno-6-13></a><span class=w>    </span><span class=nt>"skip_deseq2_qc"</span><span class=p>:</span><span class=w> </span><span class=kc>true</span><span class=p>,</span>
<a id=__codelineno-6-14 name=__codelineno-6-14 href=#__codelineno-6-14></a><span class=w>    </span><span class=nt>"skip_bbsplit"</span><span class=p>:</span><span class=w> </span><span class=kc>true</span><span class=p>,</span>
<a id=__codelineno-6-15 name=__codelineno-6-15 href=#__codelineno-6-15></a><span class=w>    </span><span class=nt>"skip_alignment"</span><span class=p>:</span><span class=w> </span><span class=kc>true</span><span class=p>,</span>
<a id=__codelineno-6-16 name=__codelineno-6-16 href=#__codelineno-6-16></a><span class=w>    </span><span class=nt>"extra_salmon_quant_args"</span><span class=p>:</span><span class=w> </span><span class=s2>"--gcBias"</span>
<a id=__codelineno-6-17 name=__codelineno-6-17 href=#__codelineno-6-17></a><span class=p>}</span>
</code></pre></div> <p><strong><code>--input</code> parameter</strong></p> <p>The <code>--input</code> parameter points to the <code>samplesheet.csv</code> file that contains all the info regarding our samples. The file looks like this:</p> <table> <thead> <tr> <th style="text-align: left;">sample</th> <th style="text-align: left;">fastq_1</th> <th style="text-align: right;">fastq_2</th> <th style="text-align: left;">strandedness</th> <th style="text-align: left;">condition</th> </tr> </thead> <tbody> <tr> <td style="text-align: left;">control_3</td> <td style="text-align: left;">778339/raw_reads/Control_3.fastq.gz</td> <td style="text-align: right;">nan</td> <td style="text-align: left;">unstranded</td> <td style="text-align: left;">control</td> </tr> <tr> <td style="text-align: left;">control_2</td> <td style="text-align: left;">778339/raw_reads/Control_2.fastq.gz</td> <td style="text-align: right;">nan</td> <td style="text-align: left;">unstranded</td> <td style="text-align: left;">control</td> </tr> <tr> <td style="text-align: left;">control_1</td> <td style="text-align: left;">778339/raw_reads/Control_1.fastq.gz</td> <td style="text-align: right;">nan</td> <td style="text-align: left;">unstranded</td> <td style="text-align: left;">control</td> </tr> <tr> <td style="text-align: left;">vampirium_3</td> <td style="text-align: left;">778339/raw_reads/Vampirium_3.fastq.gz</td> <td style="text-align: right;">nan</td> <td style="text-align: left;">unstranded</td> <td style="text-align: left;">vampirum</td> </tr> <tr> <td style="text-align: left;">vampirium_2</td> <td style="text-align: left;">778339/raw_reads/Vampirium_2.fastq.gz</td> <td style="text-align: right;">nan</td> <td style="text-align: left;">unstranded</td> <td style="text-align: left;">vampirium</td> </tr> <tr> <td style="text-align: left;">vampirium_1</td> <td style="text-align: left;">778339/raw_reads/Vampirium_1.fastq.gz</td> <td style="text-align: right;">nan</td> <td style="text-align: left;">unstranded</td> <td style="text-align: left;">vampirium</td> </tr> <tr> <td style="text-align: left;">garlicum_3</td> <td style="text-align: left;">778339/raw_reads/Garlicum_3.fastq.gz</td> <td style="text-align: right;">nan</td> <td style="text-align: left;">unstranded</td> <td style="text-align: left;">garlicum</td> </tr> <tr> <td style="text-align: left;">garlicum_2</td> <td style="text-align: left;">778339/raw_reads/Garlicum_2.fastq.gz</td> <td style="text-align: right;">nan</td> <td style="text-align: left;">unstranded</td> <td style="text-align: left;">garlicum</td> </tr> </tbody> </table> <p>As you can see, we have also provided an extra column called <code>condition</code> specifying the sample type. This will be very useful for our Differential Expression Analysis. In addition, you can also notice that we have a single-end RNAseq experiment in our hands. Finally, take a note at the fastq paths that we have provided! They all point to where the files are located inside our job!</p> <p><strong><code>--outdir</code> parameter</strong></p> <p>The <code>--outdir</code> parameter indicates where the results of the pipeline will be saved.</p> <p><strong><code>--fasta</code> parameter</strong></p> <p>Path to FASTA reference genome file.</p> <p><strong><code>--gtf</code> parameter</strong></p> <p>Path to GTF annotation file that contains genomic region information.</p> <p><strong><code>--pseudo_aligner</code> argument</strong></p> <p>The <code>--pseudo_aligner</code> argument indicates that we want to use salmon to quantify transcription levels.</p> <p>Finally, we are skipping several QC and extra steps that we did not explain in the <a href=04a_preprocessing.html>previous lesson</a>. Do not worry if you cannot manage to run the pipeline or you do not have the time, we have prepared a backup folder that contains the results from a traditional alignment + pseudoquantification for you to freely explore! (More about that <a href=#saved-results>below</a>).</p> <hr> <p>We can continue with the next argument:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a>-profile conda
</code></pre></div> <p>Unfortunately, the UCloud implementation of the nf-core pipelines do not currently allow the use of docker or singularity, which are the recommended profile options. However, UCloud has made sure that there is a working conda environment ready to use!</p> <p>Then the last couple of arguments:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a>--max_cpus 8 --max_memory 40GB
</code></pre></div> <p>These are nf-core specific arguments that indicates nextflow to only use as maximum the number of CPUs and RAM we have requested when we submitted the job. We are using 8 cores since it is what we requested in the submission page (e.g. if you submitted a job with 4 CPUs, this will be equal to 4). We are using slightly less RAM than we requested (48Gb) just in case there is a problem with memory overflow.</p> <hr> <h2 id=restarting-a-failed-run>Restarting a failed run<a class=headerlink href=#restarting-a-failed-run title="Permanent link">Â¶</a></h2> <p>When running a nf-core pipelines for the first time, you might encounter some errors, for example, one of your files has an incorrect path, or a program failed to do its job.</p> <div class="admonition failure"> <p class=admonition-title>Failure</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a>Error executing process &gt;
<a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a>Caused by:
<a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a>    Failed to create create conda environment
</code></pre></div> </div> <p><img alt src=img/04c_preprocessing_setup/conda_error.png></p> <p>Once you fix the error, it is possible to <em>resume</em> a pipeline instead of restarting the whole workflow. You can do this by adding the <code>-resume</code> argument to the nextflow command:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a>nextflow run nf-core/rnaseq -r 3.11.2 -params-file /work/778339/Scripts/nf-params.json -profile conda --max_cpus 8 --max_memory 40GBâ€‹ -resume
</code></pre></div> <h2 id=stopping-the-app>Stopping the app<a class=headerlink href=#stopping-the-app title="Permanent link">Â¶</a></h2> <p>Once the pipeline is done, go on <code>Runs</code> in uCloud and stop it from using more resources than necessary! This will help to keep the courses running for other people.</p> <p><img alt src=img/04c_preprocessing_setup/stopRun.png></p> <h2 id=saved-results>Saved results<a class=headerlink href=#saved-results title="Permanent link">Â¶</a></h2> <p>After finishing the job, everything that the pipeline has created will be saved in your own personal <em>"Jobs"</em> folder. Inside this folder there will be a subfolder called <em>nf-core: rnaseq</em>, which will contain all the jobs you have run with the nf-core app. Inside this folder, you will find the results folder named after the job name you gave when you submitted the job.</p> <ol> <li>Your material will be saved in a volume with your username, that you should be able to see under the menu <code>Files</code>.</li> </ol> <p><img alt src=img/04c_preprocessing_setup/savedWork1.png></p> <ol> <li>Go to <code>Jobs</code> -&gt; <code>nf-core: rnaseq</code> -&gt; <code>job_name</code> -&gt; <code>preprocessing_results_salmon</code></li> </ol> <p><img alt src=img/04c_preprocessing_setup/savedWork2.png></p> <p>Now you have access to the full results of your pipeline! As explained in the <a href=04b_pipelines.html>previous lesson</a>, the nf-core rnaseq workflow will create a MultiQC report summarizing most of the steps into a single and beautiful html file that is interactive and explorable. In addition, there will be a folder with the results of the individual QC steps as well as the alignment and quantification results. Take your time and check it all out!</p></div> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">May 21, 2024</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3h-2Z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">May 21, 2024</span> </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=04b_pipelines.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Nextflow and nf-core pipelines"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Nextflow and nf-core pipelines </div> </div> </a> <a href=05a_data_analysis_setup.html class="md-footer__link md-footer__link--next" aria-label="Next: Ucloud setup"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Ucloud setup </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> CC-BY-SA 4.0 license </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=assets/javascripts/bundle.081f42fc.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>