<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
    {
      "@context": "https://schema.org",
      "@id": "https://hds-sandbox.github.io/bulk_RNAseq_course/",
      "@type": "LearningResource",
      "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
      "description": "This workshop includes a tutorial on how to approach RNAseq data, starting from your sequencing reads (fastq files). Thus, the workshop only briefly touches upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, mainly for the purpose of outlining considerations in the downstream bioinformatic analysis. This workshop is based on the materials developed by members of the teaching team at the Harvard Chan Bioinformatics Core (HBC), a collection of modified tutorials from the DESeq2, R language vignettes and the nf-core rnaseq pipeline.",
      "keywords": [
        "bioinformatics, RNAseq, bulk, NGS"
      ],
      "learningResourceType": [
        "e-learning"
      ],
      "license": [
        {
          "@id": "https://creativecommons.org/licenses/by-sa/4.0/legalcode",
          "@type": "CreativeWork"
        }
      ],
      "mentions": [
        {
          "@id": "10.5281/zenodo.8046218",
          "@type": "Thing"
        },
        {
          "@id": "https://zenodo.org/record/7956067",
          "@type": "Thing"
        }
      ],
      "name": "Introduction to bulk RNAseq analysis workshop",
      "educationalLevel": "Beginner",
      "audience": "Researchers, PhDs, Postdocs, MSc"
    }
    </script><title>Count normalization with DESeq2 - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.6543a935.min.css><link rel=stylesheet href=assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/neoteroi-mkdocs.css><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script id=__analytics>function __md_analytics(){function n(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],n("js",new Date),n("config",""),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&n("event","search",{search_term:this.value})}),document$.subscribe(function(){var a=document.forms.feedback;if(void 0!==a)for(var e of a.querySelectorAll("[type=submit]"))e.addEventListener("click",function(e){e.preventDefault();var t=document.location.pathname,e=this.getAttribute("data-md-value");n("event","feedback",{page:t,data:e}),a.firstElementChild.disabled=!0;e=a.querySelector(".md-feedback__note [data-md-value='"+e+"']");e&&(e.hidden=!1)}),a.hidden=!1}),location$.subscribe(function(e){n("config","",{page_path:e.pathname})})});var e=document.createElement("script");e.async=!0,e.src="https://www.googletagmanager.com/gtag/js?id=",document.getElementById("__analytics").insertAdjacentElement("afterEnd",e)}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#normalization class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Count normalization with DESeq2 </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class=md-tabs__link> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class=md-tabs__item> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=05a_data_analysis_setup.html class=md-tabs__link> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=summary.html class=md-tabs__link> Workshop Summary </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> <span class=md-ellipsis> Course Introduction </span> </a> </li> <li class=md-nav__item> <a href=02_experimental_planning.html class=md-nav__link> <span class=md-ellipsis> Experimental planning </span> </a> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> <span class=md-ellipsis> Data Explanation </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> <span class=md-ellipsis> Read Preprocessing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> <span class=md-ellipsis> Preprocessing steps </span> </a> </li> <li class=md-nav__item> <a href=04b_pipelines.html class=md-nav__link> <span class=md-ellipsis> Nextflow and nf-core pipelines </span> </a> </li> <li class=md-nav__item> <a href=04c_preprocessing_setup.html class=md-nav__link> <span class=md-ellipsis> UCloud setup </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5 checked> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex> <span class=md-ellipsis> Analysing RNAseq data </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=true> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=05a_data_analysis_setup.html class=md-nav__link> <span class=md-ellipsis> Ucloud setup </span> </a> </li> <li class=md-nav__item> <a href=05b_count_matrix.html class=md-nav__link> <span class=md-ellipsis> Count matrix </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Normalization </span> <span class="md-nav__icon md-icon"></span> </label> <a href=05c_count_normalization.html class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Normalization </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#common-normalization-methods class=md-nav__link> <span class=md-ellipsis> Common normalization methods </span> </a> </li> <li class=md-nav__item> <a href=#rpkmfpkm-not-recommended class=md-nav__link> <span class=md-ellipsis> RPKM/FPKM (not recommended) </span> </a> </li> <li class=md-nav__item> <a href=#deseq2-normalized-counts-median-of-ratios-method class=md-nav__link> <span class=md-ellipsis> DESeq2-normalized counts: Median of ratios method </span> </a> <nav class=md-nav aria-label="DESeq2-normalized counts: Median of ratios method"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-1-creates-a-pseudo-reference-sample-row-wise-geometric-mean class=md-nav__link> <span class=md-ellipsis> Step 1: creates a pseudo-reference sample (row-wise geometric mean) </span> </a> </li> <li class=md-nav__item> <a href=#step-2-calculates-ratio-of-each-sample-to-the-reference class=md-nav__link> <span class=md-ellipsis> Step 2: calculates ratio of each sample to the reference </span> </a> </li> <li class=md-nav__item> <a href=#step-3-calculate-the-normalization-factor-for-each-sample-size-factor class=md-nav__link> <span class=md-ellipsis> Step 3: calculate the normalization factor for each sample (size factor) </span> </a> </li> <li class=md-nav__item> <a href=#step-4-calculate-the-normalized-count-values-using-the-normalization-factor class=md-nav__link> <span class=md-ellipsis> Step 4: calculate the normalized count values using the normalization factor </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#count-normalization-of-the-vampirium-dataset-using-deseq2 class=md-nav__link> <span class=md-ellipsis> Count normalization of the Vampirium dataset using DESeq2 </span> </a> <nav class=md-nav aria-label="Count normalization of the Vampirium dataset using DESeq2"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-match-the-metadata-and-counts-data class=md-nav__link> <span class=md-ellipsis> 1. Match the metadata and counts data </span> </a> </li> <li class=md-nav__item> <a href=#2-create-deseq2-object class=md-nav__link> <span class=md-ellipsis> 2. Create DESEq2 object </span> </a> <nav class=md-nav aria-label="2. Create DESEq2 object"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#pre-filtering class=md-nav__link> <span class=md-ellipsis> Pre-filtering </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#3-generate-the-normalized-counts class=md-nav__link> <span class=md-ellipsis> 3. Generate the normalized counts </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=06_exploratory_analysis.html class=md-nav__link> <span class=md-ellipsis> Exploratory Analysis </span> </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_5> <label class=md-nav__link for=__nav_5_5 id=__nav_5_5_label tabindex> <span class=md-ellipsis> Differential Expression </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> <span class=md-ellipsis> Design formula </span> </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> <span class=md-ellipsis> DESeq2 </span> </a> </li> <li class=md-nav__item> <a href=07b_hypothesis_testing.html class=md-nav__link> <span class=md-ellipsis> Hypothesis testing </span> </a> </li> <li class=md-nav__item> <a href=07c_DEA_visualization.html class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_6> <label class=md-nav__link for=__nav_5_6 id=__nav_5_6_label tabindex> <span class=md-ellipsis> Functional Analysis </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_6_label aria-expanded=false> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=08a_FA_genomic_annotation.html class=md-nav__link> <span class=md-ellipsis> Gene annotation </span> </a> </li> <li class=md-nav__item> <a href=08b_FA_overrepresentation.html class=md-nav__link> <span class=md-ellipsis> Over representation analysis </span> </a> </li> <li class=md-nav__item> <a href=08c_FA_GSEA.html class=md-nav__link> <span class=md-ellipsis> Class scoring </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=09_summarized_workflow.html class=md-nav__link> <span class=md-ellipsis> Summarized workflow </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=summary.html class=md-nav__link> <span class=md-ellipsis> Workshop Summary </span> </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#common-normalization-methods class=md-nav__link> <span class=md-ellipsis> Common normalization methods </span> </a> </li> <li class=md-nav__item> <a href=#rpkmfpkm-not-recommended class=md-nav__link> <span class=md-ellipsis> RPKM/FPKM (not recommended) </span> </a> </li> <li class=md-nav__item> <a href=#deseq2-normalized-counts-median-of-ratios-method class=md-nav__link> <span class=md-ellipsis> DESeq2-normalized counts: Median of ratios method </span> </a> <nav class=md-nav aria-label="DESeq2-normalized counts: Median of ratios method"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-1-creates-a-pseudo-reference-sample-row-wise-geometric-mean class=md-nav__link> <span class=md-ellipsis> Step 1: creates a pseudo-reference sample (row-wise geometric mean) </span> </a> </li> <li class=md-nav__item> <a href=#step-2-calculates-ratio-of-each-sample-to-the-reference class=md-nav__link> <span class=md-ellipsis> Step 2: calculates ratio of each sample to the reference </span> </a> </li> <li class=md-nav__item> <a href=#step-3-calculate-the-normalization-factor-for-each-sample-size-factor class=md-nav__link> <span class=md-ellipsis> Step 3: calculate the normalization factor for each sample (size factor) </span> </a> </li> <li class=md-nav__item> <a href=#step-4-calculate-the-normalized-count-values-using-the-normalization-factor class=md-nav__link> <span class=md-ellipsis> Step 4: calculate the normalized count values using the normalization factor </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#count-normalization-of-the-vampirium-dataset-using-deseq2 class=md-nav__link> <span class=md-ellipsis> Count normalization of the Vampirium dataset using DESeq2 </span> </a> <nav class=md-nav aria-label="Count normalization of the Vampirium dataset using DESeq2"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-match-the-metadata-and-counts-data class=md-nav__link> <span class=md-ellipsis> 1. Match the metadata and counts data </span> </a> </li> <li class=md-nav__item> <a href=#2-create-deseq2-object class=md-nav__link> <span class=md-ellipsis> 2. Create DESEq2 object </span> </a> <nav class=md-nav aria-label="2. Create DESEq2 object"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#pre-filtering class=md-nav__link> <span class=md-ellipsis> Pre-filtering </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#3-generate-the-normalized-counts class=md-nav__link> <span class=md-ellipsis> 3. Generate the normalized counts </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <div><h1 id=normalization>Normalization<a class=headerlink href=#normalization title="Permanent link">Â¶</a></h1> <p><strong>Last updated:</strong> <em><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">May 21, 2024</span></em></p> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>ðŸ•° <strong>Time Estimation:</strong> 40 minutes </p> <p>ðŸ’¬ <strong>Learning Objectives:</strong> </p> <ol> <li>Explore different types of normalization methods</li> <li>Become familiar with the <code>DESeqDataSet</code> object</li> <li>Understand how to normalize counts using DESeq2</li> </ol> </div> <p>The first step in the DE analysis workflow is count normalization, which is necessary to make accurate comparisons of gene expression between samples.</p> <p><img src=./img/05c_count_normalization/normalization_workflow.png width=1440 style="display: block; margin: auto;"></p> <p>The counts of mapped reads for each gene is proportional to the expression of RNA ("interesting") in addition to many other factors ("uninteresting"). Normalization is the process of scaling raw count values to account for the "uninteresting" factors. In this way the expression levels are more comparable between and/or within samples.</p> <p>The main factors often considered during normalization are:</p> <ul> <li><strong>Sequencing depth:</strong> Accounting for sequencing depth is necessary for comparison of gene expression between samples. In the example below, each gene appears to have doubled in expression in <em>Sample A</em> relative to <em>Sample B</em>, however this is a consequence of <em>Sample A</em> having double the sequencing depth.</li> </ul> <p><img src=./img/05c_count_normalization/library_size_bias.png width=1440 style="display: block; margin: auto;"></p> <div class="admonition note"> <p class=admonition-title>Note</p> <p>In the figure above, each red rectangle represents a read aligned to a gene. Reads connected by dashed lines connect a read spanning an intron.</p> </div> <ul> <li><strong>Gene length:</strong> Accounting for gene length is necessary for comparing expression between different genes within the same sample. In the example, <em>Gene 2</em> and <em>Gene 3</em> have similar levels of expression, but the number of reads mapped to <em>Gene 2</em> would be many more than the number mapped to <em>Gene 3</em> because <em>Gene 2</em> is longer.</li> </ul> <p><img src=./img/05c_count_normalization/length_bias.png width=1440 style="display: block; margin: auto;"></p> <ul> <li> <p><strong>GC-content</strong>: Genomic features such as GC-content may result in a read count biases, as GC-rich and GC-poor fragments are under-represented in RNAseq experiments. This under-representation is attributed to the fact that fragments with high and low GC-content are not adequately amplified in a standard high throughput sequencing protocol and, subsequently, that the fragments are difficult to align (correctly) to reference genome, i.e. less unique, repeat regions, etc. (<a href=https://academic.oup.com/nar/article/40/10/e72/2411059>Benjamini &amp; Speed, 2012</a> and <a href=https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-480>Risso et al, 2011</a>).</p> </li> <li> <p><strong>RNA composition:</strong> A few highly differentially expressed genes between samples, differences in the number of genes expressed between samples, or presence of contamination can skew some types of normalization methods. Accounting for RNA composition is recommended for accurate comparison of expression between samples, and is particularly important when performing differential expression analyses <a href=https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106>Anders &amp; Huber, 2010</a>.</p> </li> </ul> <p>In the example, if we were to divide each sample by the total number of counts to normalize, the counts would be greatly skewed by the DE gene, which takes up most of the counts for <em>Sample A</em>, but not <em>Sample B</em>. Most other genes for <em>Sample A</em> would be divided by the larger number of total counts and appear to be less expressed than those same genes in <em>Sample B</em>.</p> <p><img src=./img/05c_count_normalization/rna_composition_bias.png width=1440 style="display: block; margin: auto;"></p> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>While normalization is essential for differential expression analyses, it is also necessary for exploratory data analysis, visualization of data, and whenever you are exploring or comparing counts between or within samples.</p> </div> <h2 id=common-normalization-methods>Common normalization methods<a class=headerlink href=#common-normalization-methods title="Permanent link">Â¶</a></h2> <p>Several common normalization methods exist to account for these differences:</p> <table> <thead> <tr> <th style="text-align: left;">Normalization method</th> <th style="text-align: left;">Description</th> <th style="text-align: left;">Accounted factors</th> <th style="text-align: left;">Recommendations for use</th> </tr> </thead> <tbody> <tr> <td style="text-align: left;"><strong>CPM</strong> (counts per million)</td> <td style="text-align: left;">counts scaled by total number of reads</td> <td style="text-align: left;">sequencing depth</td> <td style="text-align: left;">gene count comparisons between replicates of the same samplegroup; <strong>NOT for within sample comparisons or DE analysis</strong></td> </tr> <tr> <td style="text-align: left;"><strong>TPM</strong> (transcripts per kilobase million)</td> <td style="text-align: left;">counts per length of transcript (kb) per million reads mapped</td> <td style="text-align: left;">sequencing depth and gene length</td> <td style="text-align: left;">gene count comparisons within a sample or between samples of the same sample group; <strong>NOT for DE analysis</strong></td> </tr> <tr> <td style="text-align: left;"><strong>RPKM/FPKM</strong> (reads/fragments per kilobase of exon per million reads/fragments mapped)</td> <td style="text-align: left;">similar to TPM</td> <td style="text-align: left;">sequencing depth and gene length</td> <td style="text-align: left;">gene count comparisons between genes within a sample; <strong>NOT for between sample comparisons or DE analysis</strong></td> </tr> <tr> <td style="text-align: left;"><a href=https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106>DESeq2â€™s <strong>median of ratios</strong></a></td> <td style="text-align: left;">counts divided by sample-specific size factors determined by median ratio of gene counts relative to geometric mean per gene</td> <td style="text-align: left;">sequencing depth and RNA composition</td> <td style="text-align: left;">gene count comparisons between samples and for <strong>DE analysis</strong>; <strong>NOT for within sample comparisons</strong></td> </tr> <tr> <td style="text-align: left;"><a href=https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25>EdgeRâ€™s <strong>trimmed mean of M values (TMM)</strong></a></td> <td style="text-align: left;">uses a weighted trimmed mean of the log expression ratios between samples</td> <td style="text-align: left;">sequencing depth, RNA composition</td> <td style="text-align: left;">gene count comparisons between samples and for <strong>DE analysis</strong>; <strong>NOT for within sample comparisons</strong></td> </tr> </tbody> </table> <h2 id=rpkmfpkm-not-recommended>RPKM/FPKM (not recommended)<a class=headerlink href=#rpkmfpkm-not-recommended title="Permanent link">Â¶</a></h2> <p>While TPM and RPKM/FPKM normalization methods both account for sequencing depth and gene length, RPKM/FPKM are not recommended. <strong>The reason is that the normalized count values output by the RPKM/FPKM method are not comparable between samples.</strong></p> <p>Using RPKM/FPKM normalization, the total number of RPKM/FPKM normalized counts for each sample will be different. Therefore, you cannot compare the normalized counts for each gene equally between samples.</p> <p><strong>RPKM-normalized counts table</strong></p> <table> <thead> <tr> <th>gene</th> <th style="text-align: center;">sampleA</th> <th style="text-align: center;">sampleB</th> </tr> </thead> <tbody> <tr> <td>XCR1</td> <td style="text-align: center;">5.5</td> <td style="text-align: center;">5.5</td> </tr> <tr> <td>WASHC1</td> <td style="text-align: center;">73.4</td> <td style="text-align: center;">21.8</td> </tr> <tr> <td>â€¦</td> <td style="text-align: center;">â€¦</td> <td style="text-align: center;">â€¦</td> </tr> <tr> <td>Total RPKM-normalized counts</td> <td style="text-align: center;">1,000,000</td> <td style="text-align: center;">1,500,000</td> </tr> </tbody> </table> <p>For example, in the table above, SampleA has a greater proportion of counts associated with XCR1 (5.5/1,000,000) than does sampleB (5.5/1,500,000) even though the RPKM count values are the same. Therefore, we cannot directly compare the counts for XCR1 (or any other gene) between sampleA and sampleB because the total number of normalized counts are different between samples.</p> <h2 id=deseq2-normalized-counts-median-of-ratios-method>DESeq2-normalized counts: Median of ratios method<a class=headerlink href=#deseq2-normalized-counts-median-of-ratios-method title="Permanent link">Â¶</a></h2> <p>Since tools for differential expression analysis are comparing the counts between sample groups for the same gene, gene length does not need to be accounted for by the tool. However, <strong>sequencing depth</strong> and <strong>RNA composition</strong> do need to be taken into account.</p> <p>To normalize for sequencing depth and RNA composition, DESeq2 uses the median of ratios method. On the user-end there is only one step, but on the back-end there are multiple steps involved, as described below.</p> <div class="admonition note note on the deseq2 workflow"> <p class=admonition-title>Note</p> <p>The steps below describe in detail some of the steps performed by DESeq2 when you run a single function to get DE genes. Basically, for a typical RNA-seq analysis, <strong>you would not run these steps individually</strong>.</p> </div> <h3 id=step-1-creates-a-pseudo-reference-sample-row-wise-geometric-mean>Step 1: creates a pseudo-reference sample (row-wise geometric mean)<a class=headerlink href=#step-1-creates-a-pseudo-reference-sample-row-wise-geometric-mean title="Permanent link">Â¶</a></h3> <p>For each gene, a pseudo-reference sample is created that is equal to the geometric mean across all samples.</p> <table> <thead> <tr> <th>gene</th> <th style="text-align: center;">sampleA</th> <th style="text-align: center;">sampleB</th> <th style="text-align: center;">pseudo-reference sample</th> </tr> </thead> <tbody> <tr> <td>EF2A</td> <td style="text-align: center;">1489</td> <td style="text-align: center;">906</td> <td style="text-align: center;">sqrt(1489 * 906) = <strong>1161.5</strong></td> </tr> <tr> <td>ABCD1</td> <td style="text-align: center;">22</td> <td style="text-align: center;">13</td> <td style="text-align: center;">sqrt(22 * 13) = <strong>17.7</strong></td> </tr> <tr> <td>â€¦</td> <td style="text-align: center;">â€¦</td> <td style="text-align: center;">â€¦</td> <td style="text-align: center;">â€¦</td> </tr> </tbody> </table> <h3 id=step-2-calculates-ratio-of-each-sample-to-the-reference>Step 2: calculates ratio of each sample to the reference<a class=headerlink href=#step-2-calculates-ratio-of-each-sample-to-the-reference title="Permanent link">Â¶</a></h3> <p>For every gene in a sample, the ratios (sample/ref) are calculated (as shown below). This is performed for each sample in the dataset. Since the majority of genes are not differentially expressed, the majority of genes in each sample should have similar ratios within the sample.</p> <table> <thead> <tr> <th>gene</th> <th style="text-align: center;">sampleA</th> <th style="text-align: center;">sampleB</th> <th style="text-align: center;">pseudo-reference sample</th> <th style="text-align: center;">ratio of sampleA/ref</th> <th style="text-align: center;">ratio of sampleB/ref</th> </tr> </thead> <tbody> <tr> <td>EF2A</td> <td style="text-align: center;">1489</td> <td style="text-align: center;">906</td> <td style="text-align: center;">1161.5</td> <td style="text-align: center;">1489/1161.5 = <strong>1.28</strong></td> <td style="text-align: center;">906/1161.5 = <strong>0.78</strong></td> </tr> <tr> <td>ABCD1</td> <td style="text-align: center;">22</td> <td style="text-align: center;">13</td> <td style="text-align: center;">16.9</td> <td style="text-align: center;">22/16.9 = <strong>1.30</strong></td> <td style="text-align: center;">13/16.9 = <strong>0.77</strong></td> </tr> <tr> <td>MEFV</td> <td style="text-align: center;">793</td> <td style="text-align: center;">410</td> <td style="text-align: center;">570.2</td> <td style="text-align: center;">793/570.2 = <strong>1.39</strong></td> <td style="text-align: center;">410/570.2 = <strong>0.72</strong></td> </tr> <tr> <td>BAG1</td> <td style="text-align: center;">76</td> <td style="text-align: center;">42</td> <td style="text-align: center;">56.5</td> <td style="text-align: center;">76/56.5 = <strong>1.35</strong></td> <td style="text-align: center;">42/56.5 = <strong>0.74</strong></td> </tr> <tr> <td>MOV10</td> <td style="text-align: center;">521</td> <td style="text-align: center;">1196</td> <td style="text-align: center;">883.7</td> <td style="text-align: center;">521/883.7 = <strong>0.590</strong></td> <td style="text-align: center;">1196/883.7 = <strong>1.35</strong></td> </tr> <tr> <td>â€¦</td> <td style="text-align: center;">â€¦</td> <td style="text-align: center;">â€¦</td> <td style="text-align: center;">â€¦</td> <td style="text-align: center;"></td> <td style="text-align: center;"></td> </tr> </tbody> </table> <h3 id=step-3-calculate-the-normalization-factor-for-each-sample-size-factor>Step 3: calculate the normalization factor for each sample (size factor)<a class=headerlink href=#step-3-calculate-the-normalization-factor-for-each-sample-size-factor title="Permanent link">Â¶</a></h3> <p>The median value (column-wise for the above table) of all ratios for a given sample is taken as the normalization factor (size factor) for that sample, as calculated below. Notice that the differentially expressed genes should not affect the median value:</p> <p><code>normalization_factor_sampleA &lt;- median(c(1.28, 1.3, 1.39, 1.35, 0.59))</code></p> <p><code>normalization_factor_sampleB &lt;- median(c(0.78, 0.77, 0.72, 0.74, 1.35))</code></p> <p>The figure below illustrates the median value for the distribution of all gene ratios for a single sample (frequency is on the y-axis).</p> <p><img src=./img/05c_count_normalization/deseq_median_of_ratios.png width=651 style="display: block; margin: auto;"></p> <p>The median of ratios method assumes that not ALL genes are differentially expressed; therefore, the normalization factors should account for sequencing depth and RNA composition of the sample (large outlier genes will not represent the median ratio values). <strong>This method is robust to imbalance in up-/down-regulation and large numbers of differentially expressed genes.</strong></p> <div class="admonition warning"> <p class=admonition-title>Warning</p> <p>Usually, these size factors are around 1, if you see large variations between samples it is important to take note since it might indicate the presence of extreme outliers.</p> </div> <h3 id=step-4-calculate-the-normalized-count-values-using-the-normalization-factor>Step 4: calculate the normalized count values using the normalization factor<a class=headerlink href=#step-4-calculate-the-normalized-count-values-using-the-normalization-factor title="Permanent link">Â¶</a></h3> <p>This is performed by dividing each raw count value in a given sample by that sample's normalization factor to generate normalized count values. This is performed for all count values (every gene in every sample). For example, if the median ratio for SampleA was 1.3 and the median ratio for SampleB was 0.77, you could calculate normalized counts as follows:</p> <p>SampleA median ratio = 1.3</p> <p>SampleB median ratio = 0.77</p> <p><strong>Raw Counts</strong></p> <table> <thead> <tr> <th>gene</th> <th style="text-align: center;">sampleA</th> <th style="text-align: center;">sampleB</th> </tr> </thead> <tbody> <tr> <td>EF2A</td> <td style="text-align: center;">1489</td> <td style="text-align: center;">906</td> </tr> <tr> <td>ABCD1</td> <td style="text-align: center;">22</td> <td style="text-align: center;">13</td> </tr> <tr> <td>â€¦</td> <td style="text-align: center;">â€¦</td> <td style="text-align: center;">â€¦</td> </tr> </tbody> </table> <p><strong>Normalized Counts</strong></p> <table> <thead> <tr> <th>gene</th> <th style="text-align: center;">sampleA</th> <th style="text-align: center;">sampleB</th> </tr> </thead> <tbody> <tr> <td>EF2A</td> <td style="text-align: center;">1489 / 1.3 = <strong>1145.39</strong></td> <td style="text-align: center;">906 / 0.77 = <strong>1176.62</strong></td> </tr> <tr> <td>ABCD1</td> <td style="text-align: center;">22 / 1.3 = <strong>16.92</strong></td> <td style="text-align: center;">13 / 0.77 = <strong>16.88</strong></td> </tr> <tr> <td>â€¦</td> <td style="text-align: center;">â€¦</td> <td style="text-align: center;">â€¦</td> </tr> </tbody> </table> <div class="admonition warning"> <p class=admonition-title>Warning</p> <p>Please note that normalized count values are not whole numbers.</p> </div> <div class="admonition question"> <p class=admonition-title><strong>Exercise 1</strong></p> <p>Determine the normalized (median of ratios) counts for your gene of interest, PD1, given the raw counts and size factors below. </p> <p>NOTE: You will need to run the code below to generate the raw counts dataframe (PD1) and the size factor vector (size_factors), then use these objects to determine the normalized counts values:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1># Raw counts for PD1</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=n>PD1</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>t</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=m>21</span><span class=p>,</span><span class=w> </span><span class=m>58</span><span class=p>,</span><span class=w> </span><span class=m>17</span><span class=p>,</span><span class=w> </span><span class=m>97</span><span class=p>,</span><span class=w> </span><span class=m>83</span><span class=p>,</span><span class=w> </span><span class=m>10</span><span class=p>))</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=nf>as_tibble</span><span class=p>()</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=nf>rename_all</span><span class=p>(</span><span class=o>~</span><span class=nf>paste0</span><span class=p>(</span><span class=s>"Sample"</span><span class=p>,</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>6</span><span class=p>))</span>
<a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a>
<a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a>
<a id=__codelineno-0-7 name=__codelineno-0-7 href=#__codelineno-0-7></a><span class=c1># Size factors for each sample</span>
<a id=__codelineno-0-8 name=__codelineno-0-8 href=#__codelineno-0-8></a><span class=n>size_factors</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=m>1.32</span><span class=p>,</span><span class=w> </span><span class=m>0.70</span><span class=p>,</span><span class=w> </span><span class=m>1.04</span><span class=p>,</span><span class=w> </span><span class=m>1.27</span><span class=p>,</span><span class=w> </span><span class=m>1.11</span><span class=p>,</span><span class=w> </span><span class=m>0.85</span><span class=p>)</span>
</code></pre></div> </div> <details class=question> <summary><strong>Solution to Exercise 1</strong></summary> <p>Let's check first what is PD1</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=n>PD1</span>
</code></pre></div> <p>Since we have the size factors per sample, we only need to divide our PD1 counts by the size factors!</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=n>PD1</span><span class=o>/</span><span class=n>size_factors</span>
</code></pre></div> </details> <h2 id=count-normalization-of-the-vampirium-dataset-using-deseq2>Count normalization of the Vampirium dataset using DESeq2<a class=headerlink href=#count-normalization-of-the-vampirium-dataset-using-deseq2 title="Permanent link">Â¶</a></h2> <p>Now that we know the theory of count normalization, we will normalize the counts for the Vampirium dataset using DESeq2. This requires a few steps:</p> <ol> <li>Ensure the row names of the metadata dataframe are present and in the same order as the column names of the counts dataframe.</li> <li>Create a <code>DESeqDataSet</code> object</li> <li>Generate the normalized counts</li> </ol> <h3 id=1-match-the-metadata-and-counts-data>1. Match the metadata and counts data<a class=headerlink href=#1-match-the-metadata-and-counts-data title="Permanent link">Â¶</a></h3> <p>We should always make sure that we have sample names that match between the two files, and that the samples are in the right order. DESeq2 will output an error if this is not the case. Since we built our <code>txi</code> object from our metadata, everything should be OK.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=c1>### Check that sample names match in both files</span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=nf>all</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>txi</span><span class=o>$</span><span class=n>counts</span><span class=p>)</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=n>meta</span><span class=o>$</span><span class=n>sample</span><span class=p>)</span>
<a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a><span class=nf>all</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>txi</span><span class=o>$</span><span class=n>counts</span><span class=p>)</span><span class=w> </span><span class=o>==</span><span class=w> </span><span class=n>meta</span><span class=o>$</span><span class=n>sample</span><span class=p>)</span>
</code></pre></div> <p>If your data did not match, you could use the <code>match()</code> function to rearrange them to be matching. <code>match()</code> function will take two arguments and find in which order the indexes of the second argument match the first argument.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=n>a</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>"a"</span><span class=p>,</span><span class=s>"b"</span><span class=p>,</span><span class=s>"c"</span><span class=p>)</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=n>b</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>"b"</span><span class=p>,</span><span class=s>"c"</span><span class=p>,</span><span class=s>"a"</span><span class=p>)</span>
<a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a>
<a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a><span class=n>reorder</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>match</span><span class=p>(</span><span class=n>a</span><span class=p>,</span><span class=n>b</span><span class=p>)</span>
<a id=__codelineno-4-5 name=__codelineno-4-5 href=#__codelineno-4-5></a><span class=n>reorder</span>
<a id=__codelineno-4-6 name=__codelineno-4-6 href=#__codelineno-4-6></a>
<a id=__codelineno-4-7 name=__codelineno-4-7 href=#__codelineno-4-7></a><span class=n>b</span><span class=p>[</span><span class=n>reorder</span><span class=p>]</span>
</code></pre></div> <div class="admonition question"> <p class=admonition-title><strong>Exercise 2</strong></p> <p>Suppose we had sample names matching in the txi object and metadata file, but they were out of order. Write the line(s) of code required make the <code>meta_random</code> dataframe with rows ordered such that they were identical to the column names of the <code>txi</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=c1># randomize metadata rownames</span>
<a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a><span class=n>meta_random</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>meta</span><span class=p>[</span><span class=nf>sample</span><span class=p>(</span><span class=m>1</span><span class=o>:</span><span class=nf>nrow</span><span class=p>(</span><span class=n>meta</span><span class=p>)),]</span>
</code></pre></div> </div> <details class=question> <summary><strong>Solution to Exercise 2</strong></summary> <p>Let's check now meta_random order:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=n>meta_random</span>
</code></pre></div> <p>We can see that it is all scrambled. We want the rows of <code>meta_random</code> to be the same order as the columns of the <code>txi@counts</code> object (which is not, as you can see below):</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1>### Check that sample names match in both files</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=nf>all</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>txi</span><span class=o>$</span><span class=n>counts</span><span class=p>)</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=n>meta_random</span><span class=o>$</span><span class=n>sample</span><span class=p>)</span><span class=w> </span><span class=c1># are all samples in our metadata?</span>
<a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a><span class=nf>all</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>txi</span><span class=o>$</span><span class=n>counts</span><span class=p>)</span><span class=w> </span><span class=o>==</span><span class=w> </span><span class=n>meta_random</span><span class=o>$</span><span class=n>sample</span><span class=p>)</span><span class=w> </span><span class=c1># are all samples in the same order?</span>
</code></pre></div> <p>Let's use the match function. First we find the order that <code>meta_random$sample</code> should be to match the columns of <code>txi@counts</code>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=n>reorder</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>match</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>txi</span><span class=o>$</span><span class=n>counts</span><span class=p>),</span><span class=n>meta_random</span><span class=o>$</span><span class=n>sample</span><span class=p>)</span>
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=n>reorder</span>
</code></pre></div> <p>Finally, we change the order of the rows of meta_random:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=n>meta_random</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>meta_random</span><span class=p>[</span><span class=n>reorder</span><span class=p>,]</span>
<a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a><span class=n>meta_random</span>
</code></pre></div> <p>And confirm:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=nf>all</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>txi</span><span class=o>$</span><span class=n>counts</span><span class=p>)</span><span class=w> </span><span class=o>==</span><span class=w> </span><span class=n>meta_random</span><span class=o>$</span><span class=n>sample</span><span class=p>)</span><span class=w> </span><span class=c1># are all samples in the same order?</span>
</code></pre></div> </details> <h3 id=2-create-deseq2-object>2. Create DESEq2 object<a class=headerlink href=#2-create-deseq2-object title="Permanent link">Â¶</a></h3> <p>Bioconductor software packages often define and use a custom class within R for storing data (input data, intermediate data and also results). These custom data structures are similar to <code>lists</code> in that they can contain multiple different data types/structures within them. But, unlike lists they have pre-specified <code>data slots</code>, which hold specific types/classes of data. The data stored in these pre-specified slots can be accessed by using specific package-defined functions.</p> <p>Let's start by creating the <code>DESeqDataSet</code> object, and then we can talk a bit more about what is stored inside it. To create the object, we will need the <strong>txi</strong> object and the <strong>metadata</strong> table as input (<code>colData</code> argument). We will also need to specify a <strong>design formula</strong>. The design formula specifies which column(s) of our metadata we want to use for statistical testing and modeling (more about that later!). For our dataset we only have one column we are interested in, which is <code>condition</code>. This column has three factor levels, which tells DESeq2 that for each gene we want to evaluate gene expression change with respect to these different levels.</p> <p><strong>It is very important to establish beforehand which sample type will be our "base" or "reference" level.</strong> If nothing is changed, DESeq2 will assume that our reference samples will be the first sample type (in alphabetical order). You can check this using the <code>factor()</code> function.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=nf>factor</span><span class=p>(</span><span class=n>meta</span><span class=o>$</span><span class=n>condition</span><span class=p>)</span>
</code></pre></div> <p>While in a normal experiment we would use control samples as our reference, in our case we are interested in both checking the differences between control vs. vampirium and garlicum vs. vampirium. Thus, it would be much more convinient to reorganize our factor base level to <code>vampirium</code>. We can do this also with the <code>factor()</code> function, using the <code>levels =</code> argument.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=n>meta</span><span class=o>$</span><span class=n>condition</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>factor</span><span class=p>(</span><span class=n>meta</span><span class=o>$</span><span class=n>condition</span><span class=p>,</span><span class=w> </span><span class=n>levels</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>"vampirium"</span><span class=p>,</span><span class=w> </span><span class=s>"control"</span><span class=p>,</span><span class=w> </span><span class=s>"garlicum"</span><span class=p>))</span>
<a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a><span class=nf>factor</span><span class=p>(</span><span class=n>meta</span><span class=o>$</span><span class=n>condition</span><span class=p>)</span>
</code></pre></div> <p>We can see now that vampirium is the first factor! Meaning that it will be interpreted by DESeq as our reference sample type.</p> <p><strong>Our count matrix input is stored in the <code>txi</code> list object</strong>. So we need to specify that using the <code>DESeqDataSetFromTximport()</code> function, which will extract the counts component and round the values to the nearest whole number.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=c1># colData argument requires rownames in order to assess matching sample names</span>
<a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=c1># meta is a tibble object from tidyverse, so we neeed to add rownames.</span>
<a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a><span class=c1># If you do not do this and the samples do not match, you will add wrong info!</span>
<a id=__codelineno-13-4 name=__codelineno-13-4 href=#__codelineno-13-4></a>
<a id=__codelineno-13-5 name=__codelineno-13-5 href=#__codelineno-13-5></a><span class=n>dds</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DESeqDataSetFromTximport</span><span class=p>(</span><span class=n>txi</span><span class=p>,</span>
<a id=__codelineno-13-6 name=__codelineno-13-6 href=#__codelineno-13-6></a><span class=w>                                   </span><span class=n>colData</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>"sample"</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-13-7 name=__codelineno-13-7 href=#__codelineno-13-7></a><span class=w>                              </span><span class=n>design</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>~</span><span class=w> </span><span class=n>condition</span><span class=p>)</span>
</code></pre></div> <details class=note> <summary>Control is not reference level warning</summary> <p>The warning from the chunk before is telling us that we have setup our vampirium samples as reference, instead of control! This is exactly what we wanted.</p> </details> <details class=note> <summary>Starting from a traditional count matrix</summary> <p>If you did not create pseudocounts, but a count matrix from aligned BAM files and tools such as <code>featurecounts</code>, you would want to use the <code>DESeqDataSetFromMatrix()</code> function.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=c1>## DO NOT RUN!</span>
<a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a><span class=c1>## Create DESeq2Dataset object from traditional count matrix</span>
<a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a><span class=n>dds</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DESeqDataSetFromMatrix</span><span class=p>(</span><span class=n>countData</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"../Data/Vampirium_counts_traditional.tsv"</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-14-4 name=__codelineno-14-4 href=#__codelineno-14-4></a><span class=w>                          </span><span class=n>colData</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>"sample"</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-14-5 name=__codelineno-14-5 href=#__codelineno-14-5></a><span class=w>                          </span><span class=n>design</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>~</span><span class=w> </span><span class=n>condition</span><span class=p>)</span>
</code></pre></div> </details> <p>You can use DESeq-specific functions to access the different slots and retrieve information, if you wish. For example, suppose we wanted the original count matrix we would use <code>counts()</code> (<em>Note: we nested it within the <code>View()</code> function so that rather than getting printed in the console we can see it in the script editor</em>) :</p> <div class=highlight><pre><span></span><code><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=nf>View</span><span class=p>(</span><span class=nf>counts</span><span class=p>(</span><span class=n>dds</span><span class=p>))</span>
</code></pre></div> <p>As we go through the workflow we will use the relevant functions to check what information gets stored inside our object.</p> <p>You can use DESeq-specific functions to access the different slots and retrieve information, if you wish. For example, suppose we wanted the original count matrix we would use <code>counts()</code>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a><span class=nf>head</span><span class=p>(</span><span class=nf>counts</span><span class=p>(</span><span class=n>dds</span><span class=p>))</span>
</code></pre></div> <p>As we go through the workflow we will use the relevant functions to check what information gets stored inside our object.</p> <h4 id=pre-filtering>Pre-filtering<a class=headerlink href=#pre-filtering title="Permanent link">Â¶</a></h4> <p>While it is not necessary to pre-filter low count genes before running the DESeq2 functions, there are two reasons which make pre-filtering useful:</p> <ul> <li>By removing rows in which there are very few reads, we reduce the memory size of the dds data object, and we increase the speed of the transformation and testing functions within DESeq2.</li> <li>It can also improve visualizations, as features with no information for differential expression are not plotted.</li> </ul> <p>Here we perform a minimal pre-filtering to keep only rows that have at least 10 reads total.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a><span class=n>keep</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>rowSums</span><span class=p>(</span><span class=nf>counts</span><span class=p>(</span><span class=n>dds</span><span class=p>))</span><span class=w> </span><span class=o>&gt;=</span><span class=w> </span><span class=m>10</span>
<a id=__codelineno-17-2 name=__codelineno-17-2 href=#__codelineno-17-2></a><span class=n>dds</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>dds</span><span class=p>[</span><span class=n>keep</span><span class=p>,]</span>
</code></pre></div> <h3 id=3-generate-the-normalized-counts>3. Generate the normalized counts<a class=headerlink href=#3-generate-the-normalized-counts title="Permanent link">Â¶</a></h3> <p>The next step is to normalize the count data in order to be able to make fair gene comparisons between samples.</p> <p>To perform the <strong>median of ratios method</strong> of normalization, DESeq2 has a single <code>estimateSizeFactors()</code> function that will generate size factors for us. We will use the function in the example below, but <strong>in a typical RNA-seq analysis this step is automatically performed by the <code>DESeq()</code> function</strong>, which we will see later.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a><span class=n>dds</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>estimateSizeFactors</span><span class=p>(</span><span class=n>dds</span><span class=p>)</span>
</code></pre></div> <p>By assigning the results back to the <code>dds</code> object we are filling in the slots of the <code>DESeqDataSet</code> object with the appropriate information. We can take a look at the normalization factor applied to each sample using:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a><span class=nf>sizeFactors</span><span class=p>(</span><span class=n>dds</span><span class=p>)</span>
</code></pre></div> <p>Now, to retrieve the normalized counts matrix from <code>dds</code>, we use the <code>counts()</code> function and add the argument <code>normalized=TRUE</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-20-1 name=__codelineno-20-1 href=#__codelineno-20-1></a><span class=n>normalized_counts</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>counts</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>normalized</span><span class=o>=</span><span class=kc>TRUE</span><span class=p>)</span>
</code></pre></div> <p>We can save this normalized data matrix to file for later use:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-21-1 name=__codelineno-21-1 href=#__codelineno-21-1></a><span class=nf>write.table</span><span class=p>(</span><span class=n>normalized_counts</span><span class=p>,</span><span class=w> </span><span class=n>file</span><span class=o>=</span><span class=s>"/work/Intro_to_bulkRNAseq/Results/normalized_counts.txt"</span><span class=p>,</span><span class=w> </span><span class=n>sep</span><span class=o>=</span><span class=s>"\t"</span><span class=p>,</span><span class=w> </span><span class=n>quote</span><span class=o>=</span><span class=bp>F</span><span class=p>)</span>
</code></pre></div> <div class="admonition warning"> <p class=admonition-title>Warning</p> <p>DESeq2 doesn't actually use normalized counts, rather it uses the raw counts and models the normalization inside the Generalized Linear Model (GLM). These normalized counts will be useful for downstream visualization of results, but cannot be used as input to DESeq2 or any other tools that perform differential expression analysis which use the negative binomial model.</p> </div> <hr> <p><em>This lesson was originally developed by members of the teaching team (Mary Piper, Meeta Mistry, Radhika Khetani) at the <a href=http://bioinformatics.sph.harvard.edu/ >Harvard Chan Bioinformatics Core (HBC)</a>.</em></p></div> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">May 21, 2024</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3h-2Z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">May 21, 2024</span> </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=05b_count_matrix.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Count matrix"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Count matrix </div> </div> </a> <a href=06_exploratory_analysis.html class="md-footer__link md-footer__link--next" aria-label="Next: Exploratory Analysis"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Exploratory Analysis </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> CC-BY-SA 4.0 license </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=assets/javascripts/bundle.081f42fc.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>