<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
    {
      "@context": "https://schema.org",
      "@id": "https://hds-sandbox.github.io/bulk_RNAseq_course/",
      "@type": "LearningResource",
      "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
      "description": "This workshop includes a tutorial on how to approach RNAseq data, starting from your sequencing reads (fastq files). Thus, the workshop only briefly touches upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, mainly for the purpose of outlining considerations in the downstream bioinformatic analysis. This workshop is based on the materials developed by members of the teaching team at the Harvard Chan Bioinformatics Core (HBC), a collection of modified tutorials from the DESeq2, R language vignettes and the nf-core rnaseq pipeline.",
      "keywords": [
        "bioinformatics, RNAseq, bulk, NGS"
      ],
      "learningResourceType": [
        "e-learning"
      ],
      "license": [
        {
          "@id": "https://creativecommons.org/licenses/by-sa/4.0/legalcode",
          "@type": "CreativeWork"
        }
      ],
      "mentions": [
        {
          "@id": "10.5281/zenodo.8046218",
          "@type": "Thing"
        },
        {
          "@id": "https://zenodo.org/record/7956067",
          "@type": "Thing"
        }
      ],
      "name": "Introduction to bulk RNAseq analysis workshop",
      "educationalLevel": "Beginner",
      "audience": "Researchers, PhDs, Postdocs, MSc"
    }
    </script><title>Log Fold Shrinkage and DEA visualizations - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.6543a935.min.css><link rel=stylesheet href=assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/neoteroi-mkdocs.css><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script id=__analytics>function __md_analytics(){function n(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],n("js",new Date),n("config",""),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&n("event","search",{search_term:this.value})}),document$.subscribe(function(){var a=document.forms.feedback;if(void 0!==a)for(var e of a.querySelectorAll("[type=submit]"))e.addEventListener("click",function(e){e.preventDefault();var t=document.location.pathname,e=this.getAttribute("data-md-value");n("event","feedback",{page:t,data:e}),a.firstElementChild.disabled=!0;e=a.querySelector(".md-feedback__note [data-md-value='"+e+"']");e&&(e.hidden=!1)}),a.hidden=!1}),location$.subscribe(function(e){n("config","",{page_path:e.pathname})})});var e=document.createElement("script");e.async=!0,e.src="https://www.googletagmanager.com/gtag/js?id=",document.getElementById("__analytics").insertAdjacentElement("afterEnd",e)}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#dea-visualization-and-log-fold-shrinkage class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Log Fold Shrinkage and DEA visualizations </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class=md-tabs__link> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class=md-tabs__item> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=05a_data_analysis_setup.html class=md-tabs__link> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=summary.html class=md-tabs__link> Workshop Summary </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> <span class=md-ellipsis> Course Introduction </span> </a> </li> <li class=md-nav__item> <a href=02_experimental_planning.html class=md-nav__link> <span class=md-ellipsis> Experimental planning </span> </a> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> <span class=md-ellipsis> Data Explanation </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> <span class=md-ellipsis> Read Preprocessing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> <span class=md-ellipsis> Preprocessing steps </span> </a> </li> <li class=md-nav__item> <a href=04b_pipelines.html class=md-nav__link> <span class=md-ellipsis> Nextflow and nf-core pipelines </span> </a> </li> <li class=md-nav__item> <a href=04c_preprocessing_setup.html class=md-nav__link> <span class=md-ellipsis> UCloud setup </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5 checked> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex> <span class=md-ellipsis> Analysing RNAseq data </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=true> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=05a_data_analysis_setup.html class=md-nav__link> <span class=md-ellipsis> Ucloud setup </span> </a> </li> <li class=md-nav__item> <a href=05b_count_matrix.html class=md-nav__link> <span class=md-ellipsis> Count matrix </span> </a> </li> <li class=md-nav__item> <a href=05c_count_normalization.html class=md-nav__link> <span class=md-ellipsis> Normalization </span> </a> </li> <li class=md-nav__item> <a href=06_exploratory_analysis.html class=md-nav__link> <span class=md-ellipsis> Exploratory Analysis </span> </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_5 checked> <label class=md-nav__link for=__nav_5_5 id=__nav_5_5_label tabindex> <span class=md-ellipsis> Differential Expression </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_5_label aria-expanded=true> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> <span class=md-ellipsis> Design formula </span> </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> <span class=md-ellipsis> DESeq2 </span> </a> </li> <li class=md-nav__item> <a href=07b_hypothesis_testing.html class=md-nav__link> <span class=md-ellipsis> Hypothesis testing </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Visualization </span> <span class="md-nav__icon md-icon"></span> </label> <a href=07c_DEA_visualization.html class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Visualization </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#visualizing-the-results class=md-nav__link> <span class=md-ellipsis> Visualizing the results </span> </a> <nav class=md-nav aria-label="Visualizing the results"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#ma-plot class=md-nav__link> <span class=md-ellipsis> MA plot </span> </a> </li> <li class=md-nav__item> <a href=#plotting-significant-de-genes class=md-nav__link> <span class=md-ellipsis> Plotting significant DE genes </span> </a> </li> <li class=md-nav__item> <a href=#heatmap class=md-nav__link> <span class=md-ellipsis> Heatmap </span> </a> </li> <li class=md-nav__item> <a href=#volcano-plot class=md-nav__link> <span class=md-ellipsis> Volcano plot </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#other-visualization-tools class=md-nav__link> <span class=md-ellipsis> Other visualization tools </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_6> <label class=md-nav__link for=__nav_5_6 id=__nav_5_6_label tabindex> <span class=md-ellipsis> Functional Analysis </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_6_label aria-expanded=false> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=08a_FA_genomic_annotation.html class=md-nav__link> <span class=md-ellipsis> Gene annotation </span> </a> </li> <li class=md-nav__item> <a href=08b_FA_overrepresentation.html class=md-nav__link> <span class=md-ellipsis> Over representation analysis </span> </a> </li> <li class=md-nav__item> <a href=08c_FA_GSEA.html class=md-nav__link> <span class=md-ellipsis> Class scoring </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=09_summarized_workflow.html class=md-nav__link> <span class=md-ellipsis> Summarized workflow </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=summary.html class=md-nav__link> <span class=md-ellipsis> Workshop Summary </span> </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#visualizing-the-results class=md-nav__link> <span class=md-ellipsis> Visualizing the results </span> </a> <nav class=md-nav aria-label="Visualizing the results"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#ma-plot class=md-nav__link> <span class=md-ellipsis> MA plot </span> </a> </li> <li class=md-nav__item> <a href=#plotting-significant-de-genes class=md-nav__link> <span class=md-ellipsis> Plotting significant DE genes </span> </a> </li> <li class=md-nav__item> <a href=#heatmap class=md-nav__link> <span class=md-ellipsis> Heatmap </span> </a> </li> <li class=md-nav__item> <a href=#volcano-plot class=md-nav__link> <span class=md-ellipsis> Volcano plot </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#other-visualization-tools class=md-nav__link> <span class=md-ellipsis> Other visualization tools </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <div><h1 id=dea-visualization-and-log-fold-shrinkage>DEA Visualization and Log Fold Shrinkage<a class=headerlink href=#dea-visualization-and-log-fold-shrinkage title="Permanent link">Â¶</a></h1> <p><strong>Last updated:</strong> <em><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">May 21, 2024</span></em></p> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>ðŸ•° <strong>Time Estimation:</strong> 75 minutes </p> <p>ðŸ’¬ <strong>Learning Objectives:</strong> </p> <ol> <li>Explain log fold change shrinkage</li> <li>Setup results data for application of visualization techniques</li> <li>Describe different data visualization useful for exploring results from a DGE analysis</li> <li>Create a volcano plot and MA plot to evaluate relationship among DGE statistics</li> <li>Create a heatmap to illustrate expression changes of differentially expressed genes</li> </ol> </div> <p>In the previous lessons, we learned about how to generate a table with Differentially Expressed genes.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1># DO NOT RUN</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=n>res_tableCont</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>results</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>contrast</span><span class=o>=</span><span class=n>contrast_cont</span><span class=p>,</span><span class=w> </span><span class=n>alpha</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.05</span><span class=p>)</span>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a>
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=nf>head</span><span class=p>(</span><span class=n>res_tableCont</span><span class=p>)</span>
</code></pre></div> <p>The problem with these fold change estimates is that they are not entirely accurate as they do not account for the large dispersion we observe with low read counts. To address this, the <strong>log2 fold changes need to be adjusted</strong>.</p> <p>To generate more accurate log2 fold change (LFC) estimates, DESeq2 allows for the <strong>shrinkage of the LFC estimates toward zero</strong> when the information for a gene is low, which could include:</p> <ul> <li>Low counts</li> <li>High dispersion values</li> </ul> <p>LFC shrinkage uses <strong>information from all genes</strong> to generate more accurate estimates. Specifically, the distribution of LFC estimates for all genes is used (as a prior) to shrink the LFC estimates of genes with little information or high dispersion toward more likely (lower) LFC estimates.</p> <p><img src=./img/07c_DEA_visualization/deseq2_shrunken_lfc.png width=483 style="display: block; margin: auto;"></p> <p><em>Illustration taken from the <a href=https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8>DESeq2 paper</a>.</em></p> <p>In the figure above, we have an example using two genes: green gene and purple gene. For each gene the expression values are plotted for each sample in the two different mouse strains (C57BL/6J and DBA/2J). Both genes have the same mean values for the two sample groups, but the green gene has little within-group variation while the purple gene has high levels of variation. For the green gene with low within-group variation, the <strong>unshrunken LFC estimate</strong> (vertex of the green <strong>solid line</strong>) is very similar to the shrunken LFC estimate (vertex of the green dotted line). However, LFC estimates for the purple gene are quite different due to the high dispersion. So even though two genes can have similar normalized count values, they can have differing degrees of LFC shrinkage. Notice the <strong>LFC estimates are shrunken toward the prior (black solid line)</strong>.</p> <p><strong>Shrinking the log2 fold changes will not change the total number of genes that are identified as significantly differentially expressed.</strong> The shrinkage of fold change is to help with downstream assessment of results. For example, if you wanted to subset your significant genes based on fold change for further evaluation, you may want to use shrunken values. Additionally, for functional analysis tools such as GSEA which require fold change values as input you would want to provide shrunken values.</p> <p>To generate the shrunken log2 fold change estimates, you have to run an additional step on your results object (that we will create below) with the function <code>lfcShrink()</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1># Save the unshrunken results to compare</span>
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a><span class=n>res_tableCont_unshrunken</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tableCont</span>
<a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a>
<a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a><span class=c1># Apply fold change shrinkage</span>
<a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a><span class=n>res_tableCont</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>lfcShrink</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>coef</span><span class=o>=</span><span class=s>"condition_control_vs_vampirium"</span><span class=p>,</span><span class=w> </span><span class=n>type</span><span class=o>=</span><span class=s>"apeglm"</span><span class=p>)</span>
</code></pre></div> <p>Depending on the version of DESeq2 you are using the default <strong>method for shrinkage estimation</strong> will differ. The defaults can be changed by adding the argument <code>type</code> in the <code>lfcShrink()</code> function as we have above. For most recent versions of DESeq2, <code>type="normal"</code> is the default and was the only method in earlier versions. It has been shown that in most situations there are alternative methods that have <a href=https://bioconductor.org/packages/devel/bioc/vignettes/apeglm/inst/doc/apeglm.html>less bias than the 'normal' method</a>, and therefore <strong>we chose to use apeglm</strong>.</p> <details class=info> <summary>More information on shrinkage</summary> <p>The DESeq2 vignette has an <a href=http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#extended-section-on-shrinkage-estimators>Extended section on shrinkage estimators</a> that is quite useful.</p> </details> <div class="admonition note"> <p class=admonition-title><code>contrast</code> vs <code>coef</code></p> <p>When using the shrinkage method, rather than using the <code>contrast</code> argument you will be required to specify <code>coef</code>. Using contrast forms an expanded model matrix, treating all factor levels equally, and averages over all distances between all pairs of factor levels to estimate the prior. Using coef, means looking only at that column of the model matrix (so usually that would be one level against the reference level) and estimates the prior for that coefficient from the distribution of those MLE of coefficients. When using coef, the shrinkage depends on which level is chosen as reference.</p> </div> <div class="admonition note"> <p class=admonition-title>How do I know what to value to provide to the <code>coef</code> argument?</p> <p>The value you provide here needs to match identically to what is stored in the column header of the coefficients table. To see what values you have to work with you can use <code>resultsNames(dds)</code>.</p> </div> <h2 id=visualizing-the-results>Visualizing the results<a class=headerlink href=#visualizing-the-results title="Permanent link">Â¶</a></h2> <p>When we are working with large amounts of data it can be useful to display that information graphically. During this lesson, we will get you started with some basic and more advanced plots commonly used when exploring differential gene expression data, however, many of these plots can be helpful in visualizing other types of data as well.</p> <p>We will be working with three different data objects we have already created in earlier lessons:</p> <ul> <li>Metadata for our samples (a dataframe): <code>meta</code></li> <li>Normalized expression data for every gene in each of our samples (a matrix): <code>normalized_counts</code></li> <li>Tibble versions of the DESeq2 results we generated in the last lesson: <code>res_tableCont_tb</code> and <code>res_tableGar_tb</code></li> </ul> <p>First, we already have a metadata tibble.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>head</span><span class=p>()</span>
</code></pre></div> <p>Next, let's bring in the <code>normalized_counts</code> object with our gene names.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=c1># DESeq2 creates a matrix when you use the counts() function</span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=c1># First convert normalized_counts to a data frame and transfer the row names to a new column called "gene"</span>
<a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a><span class=n>normalized_counts</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>counts</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>normalized</span><span class=o>=</span><span class=bp>T</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a><span class=w>  </span><span class=nf>data.frame</span><span class=p>()</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-3-5 name=__codelineno-3-5 href=#__codelineno-3-5></a><span class=w>  </span><span class=nf>rownames_to_column</span><span class=p>(</span><span class=n>var</span><span class=o>=</span><span class=s>"gene"</span><span class=p>)</span><span class=w> </span>
</code></pre></div> <h3 id=ma-plot>MA plot<a class=headerlink href=#ma-plot title="Permanent link">Â¶</a></h3> <p>A plot that can be useful to exploring our results is the MA plot. The MA plot shows the <strong>mean of the normalized counts versus the log2 fold changes for all genes tested</strong>. The genes that are significantly DE are colored to be easily identified (adjusted p-value \&lt; 0.01 by default). This is also a great way to illustrate the effect of LFC shrinkage. The DESeq2 package offers a simple function to generate an MA plot.</p> <p><strong>Let's start with the unshrunken results:</strong></p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=c1># MA plot using unshrunken fold changes</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=nf>plotMA</span><span class=p>(</span><span class=n>res_tableCont_unshrunken</span><span class=p>,</span><span class=w> </span><span class=n>ylim</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=m>-2</span><span class=p>,</span><span class=m>2</span><span class=p>))</span>
</code></pre></div> <p><strong>And now the shrunken results:</strong></p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=c1># MA plot using shrunken fold changes</span>
<a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a><span class=nf>plotMA</span><span class=p>(</span><span class=n>res_tableCont</span><span class=p>,</span><span class=w> </span><span class=n>ylim</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=m>-2</span><span class=p>,</span><span class=m>2</span><span class=p>))</span>
</code></pre></div> <p>On the left you have the unshrunken fold change values plotted and you can see the abundance of scatter for the lowly expressed genes. That is, many of these genes exhibit very high fold changes. After shrinkage, we see the fold changes are much smaller estimates.</p> <p><img src=./img/07c_DEA_visualization/maplot_unshrunken.png width=1344 style="display: block; margin: auto auto auto 0;"><img src=./img/07c_DEA_visualization/MA_plot.png width=1344 style="display: block; margin: auto auto auto 0;"></p> <p>In addition to the comparison described above, this plot allows us to evaluate the magnitude of fold changes and how they are distributed relative to mean expression. Generally, we would expect to see significant genes across the full range of expression levels.</p> <div class="admonition question"> <p class=admonition-title><strong>Exercise 1</strong></p> <p>Why are there genes with high mean and big log2 fold changes, but are not statistically significant?"</p> </div> <details class=question> <summary><strong>Solution to Exercise 1</strong></summary> <p>Because their expression values are very dispersed, and so their p-value will be very high.</p> </details> <h3 id=plotting-significant-de-genes>Plotting significant DE genes<a class=headerlink href=#plotting-significant-de-genes title="Permanent link">Â¶</a></h3> <p>One way to visualize results would be to simply plot the expression data for a handful of genes. We could do that by picking out specific genes of interest or selecting a range of genes.</p> <p><strong>Using DESeq2 <code>plotCounts()</code> to plot expression of a single gene</strong></p> <p>To pick out a specific gene of interest to plot, for example TSPAN7 (ID ENSG00000156298), we can use the <code>plotCounts()</code> from DESeq2. <code>plotCounts()</code> requires that the gene specified matches the original input to DESeq2.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=c1># Plot expression for single gene</span>
<a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=nf>plotCounts</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>gene</span><span class=o>=</span><span class=s>"ENSG00000156298"</span><span class=p>,</span><span class=w> </span><span class=n>intgroup</span><span class=o>=</span><span class=s>"condition"</span><span class=p>)</span><span class=w> </span>
</code></pre></div> <p><img src=./img/07c_DEA_visualization/countPlot.png width=1344 style="display: block; margin: auto;"></p> <div class="admonition info"> <p class=admonition-title>Info</p> <p>This DESeq2 function only allows for plotting the counts of a single gene at a time, and is not flexible regarding the appearance.</p> </div> <p><strong>Using ggplot2 to plot expression of a single gene</strong></p> <p>If you wish to change the appearance of this plot, we can save the output of <code>plotCounts()</code> to a variable specifying the <code>returnData=TRUE</code> argument, then use <code>ggplot()</code>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1># Save plotcounts to a data frame object</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=n>d</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>plotCounts</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>gene</span><span class=o>=</span><span class=s>"ENSG00000156298"</span><span class=p>,</span><span class=w> </span><span class=n>intgroup</span><span class=o>=</span><span class=s>"condition"</span><span class=p>,</span><span class=w> </span><span class=n>returnData</span><span class=o>=</span><span class=kc>TRUE</span><span class=p>)</span>
<a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a>
<a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a><span class=c1># What is the data output of plotCounts()?</span>
<a id=__codelineno-7-5 name=__codelineno-7-5 href=#__codelineno-7-5></a><span class=n>d</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>head</span><span class=p>()</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=c1># Plot the MOV10 normalized counts, using the samples (rownames(d) as labels)</span>
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>d</span><span class=p>,</span><span class=w> </span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>condition</span><span class=p>,</span><span class=w> </span><span class=n>y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>count</span><span class=p>,</span><span class=w> </span><span class=n>color</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>condition</span><span class=p>))</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a><span class=nf>geom_point</span><span class=p>(</span><span class=n>position</span><span class=o>=</span><span class=nf>position_jitter</span><span class=p>(</span><span class=n>w</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.1</span><span class=p>,</span><span class=n>h</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0</span><span class=p>))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=nf>geom_text_repel</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>label</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>rownames</span><span class=p>(</span><span class=n>d</span><span class=p>)))</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-8-5 name=__codelineno-8-5 href=#__codelineno-8-5></a><span class=nf>theme_bw</span><span class=p>()</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-8-6 name=__codelineno-8-6 href=#__codelineno-8-6></a><span class=nf>ggtitle</span><span class=p>(</span><span class=s>"TSPAN7"</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-8-7 name=__codelineno-8-7 href=#__codelineno-8-7></a><span class=nf>theme</span><span class=p>(</span><span class=n>plot.title</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>hjust</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.5</span><span class=p>))</span>
</code></pre></div> <div class="admonition note"> <p class=admonition-title>Note</p> <p>Note that in the plot below (code above), we are using <code>geom_text_repel()</code> from the <code>ggrepel</code> package to label our individual points on the plot.</p> </div> <p><img src=./img/07c_DEA_visualization/custom_countPlot.png width=1344 style="display: block; margin: auto auto auto 0;"></p> <p><strong>Create a translator from gene names to gene IDs</strong></p> <p>While gene IDs are unique and traceable, it is hard for us humans to memorize a bunch of numbers. Let's try to make a translator function that will give you possible gene IDs for a gene name. Then you can use this table to select one of the possible gene_IDs.</p> <p>The function will take as input a vector of gene names of interest, the tx2gene dataframe and the dds object that you analyzed:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=n>lookup</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>function</span><span class=p>(</span><span class=n>gene_name</span><span class=p>,</span><span class=w> </span><span class=n>tx2gene</span><span class=p>,</span><span class=w> </span><span class=n>dds</span><span class=p>){</span>
<a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a><span class=w>  </span><span class=n>hits</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>tx2gene</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>select</span><span class=p>(</span><span class=n>gene_symbol</span><span class=p>,</span><span class=w> </span><span class=n>gene_ID</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>distinct</span><span class=p>()</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a><span class=w>    </span><span class=nf>filter</span><span class=p>(</span><span class=n>gene_symbol</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=n>gene_name</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=n>gene_ID</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=nf>rownames</span><span class=p>(</span><span class=n>dds</span><span class=p>))</span>
<a id=__codelineno-9-4 name=__codelineno-9-4 href=#__codelineno-9-4></a><span class=w>  </span><span class=nf>return</span><span class=p>(</span><span class=n>hits</span><span class=p>)</span>
<a id=__codelineno-9-5 name=__codelineno-9-5 href=#__codelineno-9-5></a><span class=p>}</span>
<a id=__codelineno-9-6 name=__codelineno-9-6 href=#__codelineno-9-6></a>
<a id=__codelineno-9-7 name=__codelineno-9-7 href=#__codelineno-9-7></a><span class=nf>lookup</span><span class=p>(</span><span class=n>gene_name</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"TSPAN7"</span><span class=p>,</span><span class=w> </span><span class=n>tx2gene</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>tx2gene</span><span class=p>,</span><span class=w> </span><span class=n>dds</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>dds</span><span class=p>)</span>
</code></pre></div> <p>On the other hand, we can add the information from our tx2gene table, since it has the gene name!</p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=n>tx2gene</span>
</code></pre></div> <p>However, we see that the table has many duplicates per gene, due to the fact that a gene may have several transcripts IDs associated to it. Since our results table has gene IDs, it is important to remove transcript information and remove duplicated rows before merging the information.</p> <p>We remove the transcript ID column and duplicated rows from the tx2gene table using tidyverse syntax. We merge the tables using the <code>merge</code> function, which has many options for merging. Since our tables have different column names for the gene ID variable, we provide them with the <code>by.x</code> and <code>by.y</code> arguments. We also want to keep all of our results, so we use the argument <code>all.x</code> as well.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=n>res_tableCont_tb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>merge</span><span class=p>(</span><span class=n>res_tableCont_tb</span><span class=p>,</span><span class=w> </span><span class=n>tx2gene</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>select</span><span class=p>(</span><span class=o>-</span><span class=n>transcript_ID</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>distinct</span><span class=p>(),</span>
<a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=w>                        </span><span class=n>by.x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"gene"</span><span class=p>,</span><span class=w> </span><span class=n>by.y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"gene_ID"</span><span class=p>,</span><span class=w> </span><span class=n>all.x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>)</span>
<a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a>
<a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a><span class=n>res_tableCont_tb</span>
</code></pre></div> <h3 id=heatmap>Heatmap<a class=headerlink href=#heatmap title="Permanent link">Â¶</a></h3> <p>In addition to plotting subsets, we could also extract the normalized values of <em>all</em> the significant genes and plot a heatmap of their expression using <code>pheatmap()</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=c1># Extract normalized expression for significant genes from the vampirium and control samples</span>
<a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a><span class=c1># also get gene name</span>
<a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a><span class=n>norm_Contsig</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>normalized_counts</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>select</span><span class=p>(</span><span class=n>gene</span><span class=p>,</span><span class=w> </span><span class=nf>starts_with</span><span class=p>(</span><span class=s>"control"</span><span class=p>),</span><span class=w> </span><span class=nf>starts_with</span><span class=p>(</span><span class=s>"vampirium"</span><span class=p>))</span><span class=w> </span>
<a id=__codelineno-12-4 name=__codelineno-12-4 href=#__codelineno-12-4></a><span class=w>  </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>gene</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=n>sigCont</span><span class=o>$</span><span class=n>gene</span><span class=p>)</span><span class=w>  </span>
</code></pre></div> <p>Now let's draw the heatmap using <code>pheatmap</code>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=c1># Run pheatmap using the metadata data frame for the annotation</span>
<a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=nf>pheatmap</span><span class=p>(</span><span class=n>norm_Contsig</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>"gene"</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a><span class=w>         </span><span class=n>cluster_rows</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-13-4 name=__codelineno-13-4 href=#__codelineno-13-4></a><span class=w>         </span><span class=n>show_rownames</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>F</span><span class=p>,</span>
<a id=__codelineno-13-5 name=__codelineno-13-5 href=#__codelineno-13-5></a><span class=w>         </span><span class=n>annotation</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=n>var</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"sample"</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>select</span><span class=p>(</span><span class=s>"condition"</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-13-6 name=__codelineno-13-6 href=#__codelineno-13-6></a><span class=w>         </span><span class=n>border_color</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>NA</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-13-7 name=__codelineno-13-7 href=#__codelineno-13-7></a><span class=w>         </span><span class=n>fontsize</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>10</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-13-8 name=__codelineno-13-8 href=#__codelineno-13-8></a><span class=w>         </span><span class=n>scale</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"row"</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-13-9 name=__codelineno-13-9 href=#__codelineno-13-9></a><span class=w>         </span><span class=n>fontsize_row</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>10</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-13-10 name=__codelineno-13-10 href=#__codelineno-13-10></a><span class=w>         </span><span class=n>height</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>20</span><span class=p>)</span>
</code></pre></div> <p><img src=./img/07c_DEA_visualization/sigCont_heatmap.png width=1344 style="display: block; margin: auto auto auto 0;"></p> <div class="admonition note"> <p class=admonition-title>Note</p> <p>There are several additional arguments we have included in the function for aesthetics. One important one is <code>scale="row"</code>, in which Z-scores are plotted, rather than the actual normalized count value. </p> <p>Z-scores are computed on a gene-by-gene basis by subtracting the mean and then dividing by the standard deviation. The Z-scores are computed <strong>after the clustering</strong>, so that it only affects the graphical aesthetics and the color visualization is improved.</p> </div> <h3 id=volcano-plot>Volcano plot<a class=headerlink href=#volcano-plot title="Permanent link">Â¶</a></h3> <p>The above plot would be great to look at the expression levels of a good number of genes, but for more of a global view there are other plots we can draw. A commonly used one is a volcano plot; in which you have the log transformed adjusted p-values plotted on the y-axis and log2 fold change values on the x-axis.</p> <p>To generate a volcano plot, we first need to have a column in our results data indicating whether or not the gene is considered differentially expressed based on p-adjusted values and we will include a log2fold change here.</p> <p>To generate a volcano plot, we first need to have a column in our results data indicating whether or not the gene is considered differentially expressed based on p-adjusted values and we will include a log2fold change here.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=c1>## Obtain logical vector where TRUE values denote padj values &lt; 0.05 and fold change &gt; 1.5 in either direction</span>
<a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a>
<a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a><span class=n>res_tableCont_tb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tableCont_tb</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-14-4 name=__codelineno-14-4 href=#__codelineno-14-4></a><span class=nf>mutate</span><span class=p>(</span><span class=n>threshold_Cont</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>padj</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=m>0.05</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=nf>abs</span><span class=p>(</span><span class=n>log2FoldChange</span><span class=p>)</span><span class=w> </span><span class=o>&gt;=</span><span class=w> </span><span class=m>0.58</span><span class=p>)</span>
</code></pre></div> <p>Now we can start plotting. The <code>geom_point</code> object is most applicable, as this is essentially a scatter plot:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=c1>## Volcano plot</span>
<a id=__codelineno-15-2 name=__codelineno-15-2 href=#__codelineno-15-2></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>res_tableCont_tb</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-15-3 name=__codelineno-15-3 href=#__codelineno-15-3></a><span class=w>  </span><span class=nf>geom_point</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>log2FoldChange</span><span class=p>,</span><span class=w> </span><span class=n>y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>-</span><span class=nf>log10</span><span class=p>(</span><span class=n>padj</span><span class=p>),</span><span class=w> </span><span class=n>colour</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>threshold_Cont</span><span class=p>))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-15-4 name=__codelineno-15-4 href=#__codelineno-15-4></a><span class=w>  </span><span class=nf>ggtitle</span><span class=p>(</span><span class=s>"Control vs Vampirium"</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-15-5 name=__codelineno-15-5 href=#__codelineno-15-5></a><span class=w>  </span><span class=nf>xlab</span><span class=p>(</span><span class=s>"log2 fold change"</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-15-6 name=__codelineno-15-6 href=#__codelineno-15-6></a><span class=w>  </span><span class=nf>ylab</span><span class=p>(</span><span class=s>"-log10 adjusted p-value"</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-15-7 name=__codelineno-15-7 href=#__codelineno-15-7></a><span class=w>  </span><span class=c1>#scale_y_continuous(limits = c(0,50)) +</span>
<a id=__codelineno-15-8 name=__codelineno-15-8 href=#__codelineno-15-8></a><span class=w>  </span><span class=nf>theme</span><span class=p>(</span><span class=n>legend.position</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"none"</span><span class=p>,</span>
<a id=__codelineno-15-9 name=__codelineno-15-9 href=#__codelineno-15-9></a><span class=w>        </span><span class=n>plot.title</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>rel</span><span class=p>(</span><span class=m>1.5</span><span class=p>),</span><span class=w> </span><span class=n>hjust</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.5</span><span class=p>),</span>
<a id=__codelineno-15-10 name=__codelineno-15-10 href=#__codelineno-15-10></a><span class=w>        </span><span class=n>axis.title</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>rel</span><span class=p>(</span><span class=m>1.25</span><span class=p>)))</span><span class=w>  </span>
</code></pre></div> <p><img src=./img/07c_DEA_visualization/volcano_plot_1.png width=1344 style="display: block; margin: auto auto auto 0;"></p> <p><strong>Checking the top DE genes</strong></p> <p>This is a great way to get an overall picture of what is going on, but what if we also wanted to know where the top 10 genes (lowest padj) in our DE list are located on this plot? We could label those dots with the gene name on the Volcano plot using <code>geom_text_repel()</code>.</p> <p>First, we need to order the res_tableCont tibble by <code>padj</code>, and add an additional column to it, to include on those gene names we want to use to label the plot.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a><span class=c1>## Create an empty column to indicate which genes to label</span>
<a id=__codelineno-16-2 name=__codelineno-16-2 href=#__codelineno-16-2></a><span class=n>res_tableCont_tb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tableCont_tb</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>mutate</span><span class=p>(</span><span class=n>genelabels</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>""</span><span class=p>)</span>
<a id=__codelineno-16-3 name=__codelineno-16-3 href=#__codelineno-16-3></a>
<a id=__codelineno-16-4 name=__codelineno-16-4 href=#__codelineno-16-4></a><span class=c1>## Sort by padj values </span>
<a id=__codelineno-16-5 name=__codelineno-16-5 href=#__codelineno-16-5></a><span class=n>res_tableCont_tb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tableCont_tb</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>arrange</span><span class=p>(</span><span class=n>padj</span><span class=p>)</span>
<a id=__codelineno-16-6 name=__codelineno-16-6 href=#__codelineno-16-6></a>
<a id=__codelineno-16-7 name=__codelineno-16-7 href=#__codelineno-16-7></a><span class=c1>## Populate the gene labels column with contents of the gene symbols column for the first 10 rows, i.e. the top 10 most significantly expressed genes</span>
<a id=__codelineno-16-8 name=__codelineno-16-8 href=#__codelineno-16-8></a><span class=n>res_tableCont_tb</span><span class=o>$</span><span class=n>genelabels</span><span class=p>[</span><span class=m>1</span><span class=o>:</span><span class=m>10</span><span class=p>]</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>as.character</span><span class=p>(</span><span class=n>res_tableCont_tb</span><span class=o>$</span><span class=n>gene_symbol</span><span class=p>[</span><span class=m>1</span><span class=o>:</span><span class=m>10</span><span class=p>])</span>
<a id=__codelineno-16-9 name=__codelineno-16-9 href=#__codelineno-16-9></a>
<a id=__codelineno-16-10 name=__codelineno-16-10 href=#__codelineno-16-10></a><span class=nf>head</span><span class=p>(</span><span class=n>res_tableCont_tb</span><span class=p>)</span>
</code></pre></div> <p>Next, we plot it as before with an additional layer for <code>geom_text_repel()</code> wherein we can specify the column of gene labels we just created.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>res_tableCont_tb</span><span class=p>,</span><span class=w> </span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>log2FoldChange</span><span class=p>,</span><span class=w> </span><span class=n>y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>-</span><span class=nf>log10</span><span class=p>(</span><span class=n>padj</span><span class=p>)))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-17-2 name=__codelineno-17-2 href=#__codelineno-17-2></a><span class=w>  </span><span class=nf>geom_point</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>colour</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>threshold_Cont</span><span class=p>))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-17-3 name=__codelineno-17-3 href=#__codelineno-17-3></a><span class=w>  </span><span class=nf>geom_text_repel</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>label</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>genelabels</span><span class=p>))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-17-4 name=__codelineno-17-4 href=#__codelineno-17-4></a><span class=w>  </span><span class=nf>ggtitle</span><span class=p>(</span><span class=s>"Control vs Vampirium"</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-17-5 name=__codelineno-17-5 href=#__codelineno-17-5></a><span class=w>  </span><span class=nf>xlab</span><span class=p>(</span><span class=s>"log2 fold change"</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-17-6 name=__codelineno-17-6 href=#__codelineno-17-6></a><span class=w>  </span><span class=nf>ylab</span><span class=p>(</span><span class=s>"-log10 adjusted p-value"</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-17-7 name=__codelineno-17-7 href=#__codelineno-17-7></a><span class=w>  </span><span class=nf>theme</span><span class=p>(</span><span class=n>legend.position</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"none"</span><span class=p>,</span>
<a id=__codelineno-17-8 name=__codelineno-17-8 href=#__codelineno-17-8></a><span class=w>        </span><span class=n>plot.title</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>rel</span><span class=p>(</span><span class=m>1.5</span><span class=p>),</span><span class=w> </span><span class=n>hjust</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.5</span><span class=p>),</span>
<a id=__codelineno-17-9 name=__codelineno-17-9 href=#__codelineno-17-9></a><span class=w>        </span><span class=n>axis.title</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>rel</span><span class=p>(</span><span class=m>1.25</span><span class=p>)))</span><span class=w> </span>
</code></pre></div> <p><img src=./img/07c_DEA_visualization/volcano_plot_2.png width=1344 style="display: block; margin: auto auto auto 0;"></p> <h2 id=other-visualization-tools>Other visualization tools<a class=headerlink href=#other-visualization-tools title="Permanent link">Â¶</a></h2> <p>If you use the DESeq2 tool for differential expression analysis, the package 'DEGreport' can use the DESeq2 results output to make the top20 genes and the volcano plots generated above by writing a few lines of simple code. While you can customize the plots above, you may be interested in using the easier code. Below are examples of the code to create these plots:*</p> <div class=highlight><pre><span></span><code><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a><span class=n>DEGreport</span><span class=o>::</span><span class=nf>degPlot</span><span class=p>(</span><span class=n>dds</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>res</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>res</span><span class=p>,</span><span class=w> </span><span class=n>n</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>20</span><span class=p>,</span><span class=w> </span><span class=n>xs</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"type"</span><span class=p>,</span><span class=w> </span><span class=n>group</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"condition"</span><span class=p>)</span><span class=w> </span><span class=c1># dds object is output from DESeq2</span>
<a id=__codelineno-18-2 name=__codelineno-18-2 href=#__codelineno-18-2></a>
<a id=__codelineno-18-3 name=__codelineno-18-3 href=#__codelineno-18-3></a><span class=n>DEGreport</span><span class=o>::</span><span class=nf>degVolcano</span><span class=p>(</span>
<a id=__codelineno-18-4 name=__codelineno-18-4 href=#__codelineno-18-4></a><span class=w>    </span><span class=nf>data.frame</span><span class=p>(</span><span class=n>res</span><span class=p>[,</span><span class=nf>c</span><span class=p>(</span><span class=s>"log2fold change"</span><span class=p>,</span><span class=s>"padj"</span><span class=p>)]),</span><span class=w> </span><span class=c1># table - 2 columns</span>
<a id=__codelineno-18-5 name=__codelineno-18-5 href=#__codelineno-18-5></a><span class=w>    </span><span class=n>plot_text</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>data.frame</span><span class=p>(</span><span class=n>res</span><span class=p>[</span><span class=m>1</span><span class=o>:</span><span class=m>10</span><span class=p>,</span><span class=nf>c</span><span class=p>(</span><span class=s>"log2fold change"</span><span class=p>,</span><span class=s>"padj"</span><span class=p>,</span><span class=s>"id"</span><span class=p>)]))</span><span class=w> </span><span class=c1># table to add names</span>
<a id=__codelineno-18-6 name=__codelineno-18-6 href=#__codelineno-18-6></a>
<a id=__codelineno-18-7 name=__codelineno-18-7 href=#__codelineno-18-7></a><span class=c1># Available in the newer version for R 3.4</span>
<a id=__codelineno-18-8 name=__codelineno-18-8 href=#__codelineno-18-8></a><span class=n>DEGreport</span><span class=o>::</span><span class=nf>degPlotWide</span><span class=p>(</span><span class=n>dds</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>genes</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>row.names</span><span class=p>(</span><span class=n>res</span><span class=p>)[</span><span class=m>1</span><span class=o>:</span><span class=m>5</span><span class=p>],</span><span class=w> </span><span class=n>group</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"condition"</span><span class=p>)</span>
</code></pre></div> <div class="admonition question"> <p class=admonition-title><strong>Exercise 2</strong></p> <p>Create visualizations of the results from your DEA between Garlicum samples and Vampirium samples.</p> </div> <details class=question> <summary><strong>Solution to Exercise 2</strong></summary> <p>Our Garlicum results are saved in this table. However, we will want to use the LFC shrunken values</p> <div class=highlight><pre><span></span><code><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a><span class=c1># Normal results</span>
<a id=__codelineno-19-2 name=__codelineno-19-2 href=#__codelineno-19-2></a><span class=n>res_tableGar_unshrunken</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tableGar</span>
<a id=__codelineno-19-3 name=__codelineno-19-3 href=#__codelineno-19-3></a>
<a id=__codelineno-19-4 name=__codelineno-19-4 href=#__codelineno-19-4></a><span class=c1># Shrunken values</span>
<a id=__codelineno-19-5 name=__codelineno-19-5 href=#__codelineno-19-5></a><span class=n>res_tableGar</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>lfcShrink</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>coef</span><span class=o>=</span><span class=s>"condition_garlicum_vs_vampirium"</span><span class=p>,</span><span class=w> </span><span class=n>type</span><span class=o>=</span><span class=s>"apeglm"</span><span class=p>)</span>
</code></pre></div> <p>We can plot a MAplot:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-20-1 name=__codelineno-20-1 href=#__codelineno-20-1></a><span class=nf>plotMA</span><span class=p>(</span><span class=n>res_tableGar</span><span class=p>,</span><span class=w> </span><span class=n>ylim</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=m>-2</span><span class=p>,</span><span class=m>2</span><span class=p>))</span>
</code></pre></div> <p>And continue with a volcano plot and a heatmap. But first, let's merge our results with with our tx2gene table:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-21-1 name=__codelineno-21-1 href=#__codelineno-21-1></a><span class=n>res_tableGar_tb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>merge</span><span class=p>(</span><span class=n>res_tableGar_tb</span><span class=p>,</span><span class=w> </span><span class=n>tx2gene</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>select</span><span class=p>(</span><span class=o>-</span><span class=n>transcript_ID</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>distinct</span><span class=p>(),</span>
<a id=__codelineno-21-2 name=__codelineno-21-2 href=#__codelineno-21-2></a><span class=w>                        </span><span class=n>by.x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"gene"</span><span class=p>,</span><span class=w> </span><span class=n>by.y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"gene_ID"</span><span class=p>,</span><span class=w> </span><span class=n>all.x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>)</span>
<a id=__codelineno-21-3 name=__codelineno-21-3 href=#__codelineno-21-3></a>
<a id=__codelineno-21-4 name=__codelineno-21-4 href=#__codelineno-21-4></a><span class=n>res_tableGar_tb</span>
</code></pre></div> <p>Volcano plot with top 10 genes:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-22-1 name=__codelineno-22-1 href=#__codelineno-22-1></a><span class=c1>## Create an empty column to indicate which genes to label</span>
<a id=__codelineno-22-2 name=__codelineno-22-2 href=#__codelineno-22-2></a><span class=n>res_tableGar_tb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tableGar_tb</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>mutate</span><span class=p>(</span><span class=n>genelabels</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>""</span><span class=p>)</span>
<a id=__codelineno-22-3 name=__codelineno-22-3 href=#__codelineno-22-3></a>
<a id=__codelineno-22-4 name=__codelineno-22-4 href=#__codelineno-22-4></a><span class=c1>## Sort by padj values </span>
<a id=__codelineno-22-5 name=__codelineno-22-5 href=#__codelineno-22-5></a><span class=n>res_tableGar_tb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tableGar_tb</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>arrange</span><span class=p>(</span><span class=n>padj</span><span class=p>)</span>
<a id=__codelineno-22-6 name=__codelineno-22-6 href=#__codelineno-22-6></a>
<a id=__codelineno-22-7 name=__codelineno-22-7 href=#__codelineno-22-7></a><span class=c1>## Populate the gene labels column with contents of the gene symbols column for the first 10 rows, i.e. the top 10 most significantly expressed genes</span>
<a id=__codelineno-22-8 name=__codelineno-22-8 href=#__codelineno-22-8></a><span class=n>res_tableGar_tb</span><span class=o>$</span><span class=n>genelabels</span><span class=p>[</span><span class=m>1</span><span class=o>:</span><span class=m>10</span><span class=p>]</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>as.character</span><span class=p>(</span><span class=n>res_tableGar_tb</span><span class=o>$</span><span class=n>gene_symbol</span><span class=p>[</span><span class=m>1</span><span class=o>:</span><span class=m>10</span><span class=p>])</span>
<a id=__codelineno-22-9 name=__codelineno-22-9 href=#__codelineno-22-9></a>
<a id=__codelineno-22-10 name=__codelineno-22-10 href=#__codelineno-22-10></a><span class=nf>head</span><span class=p>(</span><span class=n>res_tableGar_tb</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-23-1 name=__codelineno-23-1 href=#__codelineno-23-1></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>res_tableGar_tb</span><span class=p>,</span><span class=w> </span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>log2FoldChange</span><span class=p>,</span><span class=w> </span><span class=n>y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>-</span><span class=nf>log10</span><span class=p>(</span><span class=n>padj</span><span class=p>)))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-23-2 name=__codelineno-23-2 href=#__codelineno-23-2></a><span class=w>  </span><span class=nf>geom_point</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>colour</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>threshold_Cont</span><span class=p>))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-23-3 name=__codelineno-23-3 href=#__codelineno-23-3></a><span class=w>  </span><span class=nf>geom_text_repel</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>label</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>genelabels</span><span class=p>))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-23-4 name=__codelineno-23-4 href=#__codelineno-23-4></a><span class=w>  </span><span class=nf>ggtitle</span><span class=p>(</span><span class=s>"Garlicum vs Vampirium"</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-23-5 name=__codelineno-23-5 href=#__codelineno-23-5></a><span class=w>  </span><span class=nf>xlab</span><span class=p>(</span><span class=s>"log2 fold change"</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-23-6 name=__codelineno-23-6 href=#__codelineno-23-6></a><span class=w>  </span><span class=nf>ylab</span><span class=p>(</span><span class=s>"-log10 adjusted p-value"</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-23-7 name=__codelineno-23-7 href=#__codelineno-23-7></a><span class=w>  </span><span class=nf>theme</span><span class=p>(</span><span class=n>legend.position</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"none"</span><span class=p>,</span>
<a id=__codelineno-23-8 name=__codelineno-23-8 href=#__codelineno-23-8></a><span class=w>        </span><span class=n>plot.title</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>rel</span><span class=p>(</span><span class=m>1.5</span><span class=p>),</span><span class=w> </span><span class=n>hjust</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.5</span><span class=p>),</span>
<a id=__codelineno-23-9 name=__codelineno-23-9 href=#__codelineno-23-9></a><span class=w>        </span><span class=n>axis.title</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>rel</span><span class=p>(</span><span class=m>1.25</span><span class=p>)))</span><span class=w> </span>
</code></pre></div> <p>Heatmap of DE genes for Vampirium vs Garlicum:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-24-1 name=__codelineno-24-1 href=#__codelineno-24-1></a><span class=c1># Extract normalized expression for significant genes from the vampirium and garlicum samples</span>
<a id=__codelineno-24-2 name=__codelineno-24-2 href=#__codelineno-24-2></a><span class=c1># also get gene name</span>
<a id=__codelineno-24-3 name=__codelineno-24-3 href=#__codelineno-24-3></a><span class=n>norm_Garsig</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>normalized_counts</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>select</span><span class=p>(</span><span class=n>gene</span><span class=p>,</span><span class=w> </span><span class=nf>starts_with</span><span class=p>(</span><span class=s>"garlicum"</span><span class=p>),</span><span class=w> </span><span class=nf>starts_with</span><span class=p>(</span><span class=s>"vampirium"</span><span class=p>))</span><span class=w> </span>
<a id=__codelineno-24-4 name=__codelineno-24-4 href=#__codelineno-24-4></a><span class=w>  </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>gene</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=n>sigGar</span><span class=o>$</span><span class=n>gene</span><span class=p>)</span><span class=w>  </span>
</code></pre></div> <p>Now let's draw the heatmap using <code>pheatmap</code>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-25-1 name=__codelineno-25-1 href=#__codelineno-25-1></a><span class=c1># Run pheatmap using the metadata data frame for the annotation</span>
<a id=__codelineno-25-2 name=__codelineno-25-2 href=#__codelineno-25-2></a><span class=nf>pheatmap</span><span class=p>(</span><span class=n>norm_Garsig</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>"gene"</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-25-3 name=__codelineno-25-3 href=#__codelineno-25-3></a><span class=w>         </span><span class=n>cluster_rows</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-25-4 name=__codelineno-25-4 href=#__codelineno-25-4></a><span class=w>         </span><span class=n>show_rownames</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>F</span><span class=p>,</span>
<a id=__codelineno-25-5 name=__codelineno-25-5 href=#__codelineno-25-5></a><span class=w>         </span><span class=n>annotation</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=n>var</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"sample"</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>select</span><span class=p>(</span><span class=s>"condition"</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-25-6 name=__codelineno-25-6 href=#__codelineno-25-6></a><span class=w>         </span><span class=n>border_color</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>NA</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-25-7 name=__codelineno-25-7 href=#__codelineno-25-7></a><span class=w>         </span><span class=n>fontsize</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>10</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-25-8 name=__codelineno-25-8 href=#__codelineno-25-8></a><span class=w>         </span><span class=n>scale</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"row"</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-25-9 name=__codelineno-25-9 href=#__codelineno-25-9></a><span class=w>         </span><span class=n>fontsize_row</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>10</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-25-10 name=__codelineno-25-10 href=#__codelineno-25-10></a><span class=w>         </span><span class=n>height</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>20</span><span class=p>)</span>
</code></pre></div> </details> <hr> <p><em>This lesson was originally developed by members of the teaching team (Mary Piper, Meeta Mistry, Radhika Khetani) at the <a href=http://bioinformatics.sph.harvard.edu/ >Harvard Chan Bioinformatics Core (HBC)</a>.</em></p> <p><em>Materials and hands-on activities were adapted from <a href=http://www.bioconductor.org/help/workflows/rnaseqGene/#de>RNA-seq workflow</a> on the Bioconductor website</em></p></div> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">May 21, 2024</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3h-2Z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">May 21, 2024</span> </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=07b_hypothesis_testing.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Hypothesis testing"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Hypothesis testing </div> </div> </a> <a href=08a_FA_genomic_annotation.html class="md-footer__link md-footer__link--next" aria-label="Next: Gene annotation"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Gene annotation </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> CC-BY-SA 4.0 license </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=assets/javascripts/bundle.081f42fc.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>