---
title: "Log Fold Shrinkage and DEA visualizations"
author: "You!"
date: '`r Sys.Date()`'
format:
  html:
    toc: true
    toc_float: true
    code-fold: false
    code-overflow: wrap
    df-print: paged
---

```{r knitr, include = FALSE}
DOCNAME = knitr::current_input()
DOCNAME = gsub(DOCNAME, pattern = ".Rmd", replacement = "", fixed = T)
knitr::opts_chunk$set(autodep        = TRUE,
                      cache          = FALSE,
                      echo           = TRUE,
                      error          = FALSE,
                      fig.align      = "center",
                      fig.path       = paste0("./img/", DOCNAME, "/"), #images will be put in this folder, under the notebook name
                      message        = FALSE,
                      warning        = FALSE,
                      eval           = TRUE)
```

```{r, include = FALSE, echo = FALSE}
# DO NOT RUN IF YOU HAVE ALREADY RUN PREVIOUS NOTEBOOKS
# This chunk is ONLY necessary if you want to knit this document into a pdf!!
library(tidyverse)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(DEGreport)
library(ggrepel)
library(tximport)

meta <- read_csv("../Data/samplesheet.csv")
meta$condition = factor(meta$condition, levels = c("vampirium", "control", "garlicum"))
dir <- "../Data/salmon"
tx2gene <- read_table("../Data/salmon/salmon_tx2gene.tsv", col_names = c("transcript_ID","gene_ID","gene_symbol"))
files <- file.path(dir, meta$sample, "quant.sf")
names(files) <- meta$sample
txi <- tximport(files, type="salmon", tx2gene=tx2gene, countsFromAbundance = "lengthScaledTPM", ignoreTxVersion	= TRUE)
dds <- DESeqDataSetFromTximport(txi,
                                   colData = meta %>% column_to_rownames("sample"), 
                              design = ~ condition)
keep <- rowSums(counts(dds)) > 0
dds <- dds[keep,]

dds <- DESeq(dds)

contrast_cont <- c("condition", "control","vampirium")
res_tableCont <- results(dds, contrast=contrast_cont, alpha = 0.05)
res_tableCont_tb <- res_tableCont %>%
  as_tibble(rownames = "gene") %>%
  relocate(gene, .before = baseMean)

sigCont <- res_tableCont_tb %>%
  dplyr::filter(padj < 0.05)
```

Approximate time: 45 minutes

## Learning Objectives

-   Explain log fold change shrinkage
-   Setup results data for application of visualization techniques
-   Describe different data visualization useful for exploring results from a DGE analysis
-   Create a volcano plot and MA plot to evaluate relationship among DGE statistics
-   Create a heatmap to illustrate expression changes of differentially expressed genes

## More accurate LFC estimates

In the previous lessons, we learned about how to generate a table with Differentially Expressed Genes

```{r, eval = FALSE}
## DO NOT RUN
res_tableCont <- results(dds, contrast=contrast_cont, alpha = 0.05)

head(res_tableCont)
```

The problem with these fold change estimates is that they are not entirely accurate as they do not account for the large dispersion we observe with low read counts. To address this, the **log2 fold changes need to be adjusted**.

To generate the shrunken log2 fold change estimates, you have to run an additional step on your results object (that we will create below) with the function `lfcShrink()`.

```{r}
## Save the unshrunken results to compare
res_tableCont_unshrunken <- res_tableCont

# Apply fold change shrinkage
res_tableCont <- lfcShrink(dds, coef="condition_control_vs_vampirium", type="apeglm")
```

> For more information on shrinkage, the DESeq2 vignette has an [Extended section on shrinkage estimators](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#extended-section-on-shrinkage-estimators) that is quite useful.

> #### `contrast` vs `coef`
>
> When using the shrinkage method, rather than using the `contrast` argument you will be required to specify `coef`. Using contrast forms an expanded model matrix, treating all factor levels equally, and averages over all distances between all pairs of factor levels to estimate the prior. Using coef, means looking only at that column of the model matrix (so usually that would be one level against the reference level) and estimates the prior for that coefficient from the distribution of those MLE of coefficients. When using coef, the shrinkage depends on which level is chosen as reference.

> #### How do I know what to value to provide to the `coef` argument?
>
> The value you provide here needs to match identically to what is stored in the column header of the coefficients table. To see what values you have to work with you can use `resultsNames(dds)`.

## Visualizing the results

### MA plot

A plot that can be useful to exploring our results is the MA plot. The MA plot shows the **mean of the normalized counts versus the log2 fold changes for all genes tested**. The genes that are significantly DE are colored to be easily identified (adjusted p-value < 0.01 by default).

**Let's start with the unshrunken results:**

```{r MAplot_unshrunken}
# MA plot using unshrunken fold changes
plotMA(res_tableCont_unshrunken, ylim=c(-2,2))
```

**And now the shrunken results:**

```{r MAplot_shrunken}
# MA plot using shrunken fold changes
plotMA(res_tableCont, ylim=c(-2,2))
```

***

**Exercise 1**

Why are there genes with high mean and big log2 fold changes, but are not statistically significant?

***

## Advanced visualizations

We will be working with three different data objects we have already created in earlier lessons:

-   Metadata for our samples (a dataframe): `meta`
-   Normalized expression data for every gene in each of our samples (a matrix): `normalized_counts`
-   Tibble versions of the DESeq2 results we generated in the last lesson: `res_tableCont_tb` and `res_tableGar_tb`

First, we already have a metadata tibble.
```{r}
meta %>% head()
```

Next, let's bring in the `normalized_counts` object with our gene names.

```{r}
# DESeq2 creates a matrix when you use the counts() function
## First convert normalized_counts to a data frame and transfer the row names to a new column called "gene"
normalized_counts <- counts(dds, normalized=T) %>% 
  data.frame() %>%
  rownames_to_column(var="gene") 
```

### Plotting significant DE genes

One way to visualize results would be to simply plot the expression data for a handful of genes. We could do that by picking out specific genes of interest or selecting a range of genes.

**Using DESeq2 `plotCounts()` to plot expression of a single gene**

To pick out a specific gene of interest to plot, for example TSPAN7 (ID ENSG00000156298), we can use the `plotCounts()` from DESeq2. `plotCounts()` requires that the gene specified matches the original input to DESeq2.

```{r countPlot}
# Plot expression for single gene
plotCounts(dds, gene="ENSG00000156298", intgroup="condition") 
```

**Using ggplot2 to plot expression of a single gene**

If you wish to change the appearance of this plot, we can save the output of `plotCounts()` to a variable specifying the `returnData=TRUE` argument, then use `ggplot()`:

```{r}
# Save plotcounts to a data frame object
d <- plotCounts(dds, gene="ENSG00000156298", intgroup="condition", returnData=TRUE)

# What is the data output of plotCounts()?
d %>% head()
```

```{r custom_countPlot}
# Plot the Vampirium normalized counts, using the samples (rownames(d) as labels)
library(ggrepel)

ggplot(d, aes(x = condition, y = count, color = condition)) + 
geom_point(position=position_jitter(w = 0.1,h = 0)) +
geom_text_repel(aes(label = rownames(d))) + 
theme_bw() +
ggtitle("TSPAN7") +
theme(plot.title = element_text(hjust = 0.5))
```

**Create a translator from gene names to gene IDs**

While gene IDs are unique and traceable, it is hard for us humans to memorize a bunch of numbers. Let's try to make a translator function that will give you possible gene IDs for a gene name. Then you can use this table to select one of the possible gene_IDs.

The function will take as input a vector of gene names of interest, the tx2gene dataframe and the dds object that you analyzed
```{r}
lookup <- function(gene_name, tx2gene, dds){
  hits <- tx2gene %>% dplyr::select(gene_symbol, gene_ID) %>% distinct() %>% 
    dplyr::filter(gene_symbol %in% gene_name & gene_ID %in% rownames(dds))
  return(hits)
}

lookup(gene_name = "TSPAN7", tx2gene = tx2gene, dds = dds)
```

On the other hand, we can add the information from our tx2gene table, since it has the gene name!

```{r}
tx2gene
```

However, we see that the table has many duplicates per gene, due to the fact that a gene may have several transcripts IDs associated to it. Since our results table has gene IDs, it is important to remove transcript information and remove duplicated rows before merging the information.

We remove the transcript ID column and duplicated rows from the tx2gene table using tidyverse syntax. We merge the tables using the `merge` function, which has many options for merging. Since our tables have different column names for the gene ID variable, we provide them with the `by.x` and `by.y` arguments. We also want to keep all of our results, so we use the argument `all.x` as well.

```{r}
res_tableCont_tb <- merge(res_tableCont_tb, tx2gene %>% select(-transcript_ID) %>% distinct(),
                        by.x = "gene", by.y = "gene_ID", all.x = T)

res_tableCont_tb
```

### Heatmap

In addition to plotting subsets, we could also extract the normalized values of *all* the significant genes and plot a heatmap of their expression using `pheatmap()`.

```{r}
### Extract normalized expression for significant genes from the OE and control samples
### also get gene name
norm_Contsig <- normalized_counts %>% dplyr::select(gene, starts_with("Vampirium"), starts_with("Control"))  %>%
  dplyr::filter(gene %in% sigCont$gene)  
```

Now let's draw the heatmap using `pheatmap`:

```{r sigCont_heatmap}
### Run pheatmap using the metadata data frame for the annotation
pheatmap(norm_Contsig %>% column_to_rownames("gene"), 
         cluster_rows = T, 
         show_rownames = F,
         annotation = meta %>% column_to_rownames(var = "sample") %>% dplyr::select("condition"), 
         border_color = NA, 
         fontsize = 10, 
         scale = "row", 
         fontsize_row = 10, 
         height = 20)
```

> *NOTE:* There are several additional arguments we have included in the function for aesthetics. One important one is `scale="row"`, in which Z-scores are plotted, rather than the actual normalized count value.
>
> Z-scores are computed on a gene-by-gene basis by subtracting the mean and then dividing by the standard deviation. The Z-scores are computed **after the clustering**, so that it only affects the graphical aesthetics and the color visualization is improved.

### Volcano plot

To generate a volcano plot, we first need to have a column in our results data indicating whether or not the gene is considered differentially expressed based on p-adjusted values and we will include a log2fold change here.

```{r}
## Obtain logical vector where TRUE values denote padj values < 0.05 and fold change > 1.5 in either direction

res_tableCont_tb <- res_tableCont_tb %>% 
mutate(threshold_OE = padj < 0.05 & abs(log2FoldChange) >= 0.58)
```

Now we can start plotting. The `geom_point` object is most applicable, as this is essentially a scatter plot:

```{r volcano_plot}
## Volcano plot
ggplot(res_tableCont_tb) + 
  geom_point(aes(x = log2FoldChange, y = -log10(padj), colour = threshold_OE)) +
  ggtitle("Control vs Vampirium") +
  xlab("log2 fold change") + 
  ylab("-log10 adjusted p-value") +
  #scale_y_continuous(limits = c(0,50)) +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.5), hjust = 0.5),
        axis.title = element_text(size = rel(1.25)))  
```

**Checking the top DE genes**

This is a great way to get an overall picture of what is going on, but what if we also wanted to know where the top 10 genes (lowest padj) in our DE list are located on this plot? We could label those dots with the gene name on the Volcano plot using `geom_text_repel()`.

First, we need to order the res_tableCont tibble by `padj`, and add an additional column to it, to include on those gene names we want to use to label the plot.

```{r}
## Create an empty column to indicate which genes to label
res_tableCont_tb <- res_tableCont_tb %>% mutate(genelabels = "")

## Sort by padj values 
res_tableCont_tb <- res_tableCont_tb %>% arrange(padj)

## Populate the gene labels column with contents of the gene symbols column for the first 10 rows, i.e. the top 10 most significantly expressed genes
res_tableCont_tb$genelabels[1:10] <- as.character(res_tableCont_tb$gene_symbol[1:10])

head(res_tableCont_tb)
```

Next, we plot it as before with an additional layer for `geom_text_repel()` wherein we can specify the column of gene labels we just created.

```{r volcano_plot_label}
ggplot(res_tableCont_tb, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(colour = threshold_OE)) +
  geom_text_repel(aes(label = genelabels)) +
  ggtitle("Control vs Vampirium") +
  xlab("log2 fold change") + 
  ylab("-log10 adjusted p-value") +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.5), hjust = 0.5),
        axis.title = element_text(size = rel(1.25))) 
```

------------------------------------------------------------------------

## Extra: Reports for DEA

***NOTE:** If using the DESeq2 tool for differential expression analysis, the package 'DEGreport' can use the DESeq2 results output to make the top20 genes and the volcano plots generated above by writing a few lines of simple code. While you can customize the plots above, you may be interested in using the easier code. Below are examples of the code to create these plots:*

Note that, since the argument these reports is the dds object, which contains gene IDs instead of gene names, we will have to provide gene IDs to fetch information!

```{r degPlot, fig.height=10}
DEGreport::degPlot(dds = dds, res = res_tableCont_unshrunken, n = 20, xs = "condition", group = "condition") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))# dds object is output from DESeq2
```

```{r degPlotWide}
DEGreport::degPlotWide(counts = dds, genes = row.names(res_tableCont)[1:5], group = "condition")
```

```{r degVolcano}
DEGreport::degVolcano(
    data.frame(res_tableCont[,c("log2FoldChange","padj")]), # table - 2 columns
    plot_text = data.frame(res_tableCont_tb[1:10,c("log2FoldChange","padj","gene_symbol")])) # table to add names
```
