---
title: "Functional Analysis for RNA-seq"
author: "You!"
date: '`r Sys.Date()`'
knit: (function(inputFile, encoding) { 
      rmarkdown::render(inputFile,
                        encoding=encoding,
                        output_format='all',
                        output_dir='./')})
output:
  # To create PDF report, uncomment below
  #pdf_document:
  #  toc: yes  
  html_document:
    theme: yeti # nice theme for the webpage
    toc: yes # table of contents
    toc_float: yes # table of contents "floats" in the document
    df_print: paged # data frames are interactive
    dev: png # what format do you want for the figures?
---

```{r knitr, include = FALSE}
DOCNAME = knitr::current_input()
DOCNAME = gsub(DOCNAME, pattern = ".Rmd", replacement = "", fixed = T)
knitr::opts_chunk$set(autodep        = TRUE,
                      cache          = FALSE,
                      echo           = TRUE,
                      error          = FALSE,
                      fig.align      = "center",
                      fig.path       = paste0("./img/", DOCNAME, "/"), #images will be put in this folder, under the notebook name
                      message        = FALSE,
                      warning        = FALSE,
                      eval           = TRUE)
```

```{r, include = FALSE, echo = FALSE}
# DO NOT RUN IF YOU HAVE ALREADY RUN PREVIOUS NOTEBOOKS
# This chunk is ONLY necessary if you want to knit this document into a pdf!!
library(tidyverse)
library(DESeq2)
library(ggrepel)
library(annotables)
library(tximport)

setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
meta <- read_csv("/work/Intro_to_bulkRNAseq/Data/samplesheet.csv")
dir <- "/work/Intro_to_bulkRNAseq/Data/salmon"
tx2gene <- read_table("/work/Intro_to_bulkRNAseq/Data/salmon_tx2gene.tsv", col_names = c("transcript_ID","gene_ID","gene_symbol"))
files <- file.path(dir, meta$sample, "quant.sf")
names(files) <- meta$sample
txi <- tximport(files, type="salmon", tx2gene=tx2gene, countsFromAbundance = "lengthScaledTPM", ignoreTxVersion	= TRUE)
dds <- DESeqDataSetFromTximport(txi,
                                   colData = meta %>% column_to_rownames("sample"), 
                              design = ~ condition)
keep <- rowSums(counts(dds)) > 0
dds <- dds[keep,]


dds <- DESeq(dds)

res_tableOE <- lfcShrink(dds, coef = "condition_MOV10_overexpression_vs_control")
res_tableOE_tb <- res_tableOE %>%
    data.frame() %>%
    rownames_to_column(var="gene") %>% 
    as_tibble()

ids <- grch37 %>% dplyr::filter(ensgene %in% rownames(res_tableOE)) 
res_ids <- inner_join(res_tableOE_tb, ids, by=c("gene"="ensgene"))
```

Approximate time: 45 minutes

## Learning Objectives:

*  Identify popular functional analysis tools for over-representation analysis

## clusterProfiler

We will be using [clusterProfiler](http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) to perform over-representation analysis on GO terms associated with our list of significant genes. The tool takes as input a significant gene list and a background gene list and performs statistical enrichment analysis using hypergeometric testing. The basic arguments allow the user to select the appropriate organism and GO ontology (BP, CC, MF) to test. 

### Running clusterProfiler

To run clusterProfiler GO over-representation analysis, we wil use the Ensembl IDs, instead of the gene names, since the tool works a bit easier with the Ensembl IDs. 

Then load the following libraries:

```{r}
# Load libraries
library(DOSE)
library(pathview)
library(clusterProfiler)
library(org.Hs.eg.db)
```

To perform the over-representation analysis, we need a list of background genes and a list of significant genes. For our background dataset we will use all genes tested for differential expression (all genes in our results table). For our significant gene list we will use genes with p-adjusted values less than 0.05 (we could include a fold change threshold too if we have many DE genes).

```{r}
## Create background dataset for hypergeometric testing using all genes tested for significance in the results
allOE_genes <- dplyr::filter(res_ids, !is.na(gene)) %>% 
  pull(gene) %>% 
  as.character()

## Extract significant results
sigOE <- dplyr::filter(res_ids, padj < 0.05 & !is.na(gene))

sigOE_genes <- sigOE %>% 
  pull(gene) %>% 
  as.character()
```

Now we can perform the GO enrichment analysis and save the results:

```{r}
## Run GO enrichment analysis 
ego <- enrichGO(gene = sigOE_genes, 
                universe = allOE_genes,
                keyType = "ENSEMBL",
                OrgDb = org.Hs.eg.db, 
                ont = "BP", 
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)
```

Let's check the results:

```{r}
## Output results from GO analysis to a table
cluster_summary <- data.frame(ego)
cluster_summary

write.csv(cluster_summary, "/work/Intro_to_bulkRNAseq/Results/clusterProfiler_Mov10oe.csv")
```

***

**Exercise 1**

Create two new GO enrichment analyses one with UP and another for DOWN regulated genes for MOV10 overexpression.

***

### Visualizing clusterProfiler results
clusterProfiler has a variety of options for viewing the over-represented GO terms. We will explore the dotplot, enrichment plot, and the category netplot.

The **dotplot** shows the number of genes associated with the first 50 terms (size) and the p-adjusted values for these terms (color). This plot displays the top 50 GO terms by gene ratio (# genes related to GO term / total number of sig genes), not p-adjusted value.

```{r enrichGO_dotplot, fig.height= 10, fig.width=5}
## Dotplot 
dotplot(ego, showCategory=50)
```

The next plot is the **enrichment GO plot**, which shows the relationship between the top 50 most significantly enriched GO terms (padj.), by grouping similar terms together. Before creating the plot, we will need to obtain the similarity between terms using the `pairwise_termsim()` function ([instructions for emapplot](https://rdrr.io/github/GuangchuangYu/enrichplot/man/emapplot.html)). In the enrichment plot, the color represents the p-values relative to the other displayed terms (brighter red is more significant), and the size of the terms represents the number of genes that are significant from our list.

```{r}
## Add similarity matrix to the termsim slot of enrichment result
ego <- enrichplot::pairwise_termsim(ego)
```

```{r enrichGO_emmapplot, fig.height= 10}
## Enrichmap clusters the 50 most significant (by padj) GO terms to visualize relationships between terms
emapplot(ego, showCategory = 50)
```

Finally, the **category netplot** shows the relationships between the genes associated with the top five most significant GO terms and the fold changes of the significant genes associated with these terms (color). The size of the GO terms reflects the pvalues of the terms, with the more significant terms being larger. This plot is particularly useful for hypothesis generation in identifying genes that may be important to several of the most affected processes. 

```{r}
## To color genes by log2 fold changes, we need to extract the log2 fold changes from our results table creating a named vector
OE_foldchanges <- sigOE$log2FoldChange

names(OE_foldchanges) <- sigOE$gene
```

```{r cnetplot}
## Cnetplot details the genes associated with one or more terms - by default gives the top 5 significant terms (by padj)
cnetplot(ego, 
         categorySize="pvalue", 
         showCategory = 5, 
         foldChange=OE_foldchanges, 
         vertex.label.font=6)
```

```{r}
## If some of the high fold changes are getting drowned out due to a large range, you could set a maximum fold change value
OE_foldchanges <- ifelse(OE_foldchanges > 2, 2, OE_foldchanges)
OE_foldchanges <- ifelse(OE_foldchanges < -2, -2, OE_foldchanges)
```

```{r cnetolot_limits}
cnetplot(ego, 
         categorySize="pvalue", 
         showCategory = 5, 
         foldChange=OE_foldchanges, 
         vertex.label.font=6)
```

If you are interested in significant processes that are **not** among the top five, you can subset your `ego` dataset to only display these processes:

```{r cnetplot_subset}
## Subsetting the ego results without overwriting original `ego` variable
ego2 <- ego

ego2@result <- ego@result[c(1,3,4,8,9),]

## Plotting terms of interest
cnetplot(ego2, 
         categorySize="pvalue", 
         foldChange=OE_foldchanges, 
         showCategory = 5, 
         vertex.label.font=6)
```

***

**Exercise 2**

Run a Disease Ontology (DO) overrepresentation analysis using the `enrichDO()` function. **NOTE** the arguments are very similar to the previous examples. 

- Do you find anything interesting?

***