<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="This course is an introduction for how to approach bulk RNAseq data, starting from the sequencing reads.  It will provide an overview of the fundamentals of RNAseq analysis, including read preprocessing, data normalization, data exploration with PCAs and heatmaps, performing differential expression analysis and annotation of the differentially expressed genes.  Participants will also learn how to evaluate confounding and batch effects in the data.  The course will further touch upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, especially about how they influence downstream bioinformatic analysis.   "><meta name=author content="Jose Alejandro Romero Herrera"><link rel=prev href=04c_preprocessing_setup.html><link rel=next href=05b_count_matrix.html><link rel=icon href=img/favicon.png><meta name=generator content="mkdocs-1.4.2, mkdocs-material-9.0.8"><title>Ucloud setup for data analysis - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.0d440cfe.min.css><link rel=stylesheet href=assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#setup-for-teaching-in-ucloud class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Ucloud setup for data analysis </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class=md-tabs__link> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class=md-tabs__item> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class=md-tabs__item> <a href=05a_data_analysis_setup.html class="md-tabs__link md-tabs__link--active"> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=09_summarized_workflow.html class=md-tabs__link> Summarized workflow </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> Course Introduction </a> </li> <li class=md-nav__item> <a href=02_experimental_planning.html class=md-nav__link> Experimental planning </a> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> Data Explanation </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> Read Preprocessing <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> Preprocessing steps </a> </li> <li class=md-nav__item> <a href=04b_pipelines.html class=md-nav__link> Nextflow and nf-core pipelines </a> </li> <li class=md-nav__item> <a href=04c_preprocessing_setup.html class=md-nav__link> UCloud setup </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5 checked> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> Analysing RNAseq data <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=true> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Ucloud setup <span class="md-nav__icon md-icon"></span> </label> <a href=05a_data_analysis_setup.html class="md-nav__link md-nav__link--active"> Ucloud setup </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#submit-the-job-in-ucloud class=md-nav__link> Submit the job in Ucloud </a> </li> <li class=md-nav__item> <a href=#stopping-the-app class=md-nav__link> Stopping the app </a> </li> <li class=md-nav__item> <a href=#saved-work class=md-nav__link> Saved work </a> </li> <li class=md-nav__item> <a href=#restarting-the-rstudio-session class=md-nav__link> Restarting the Rstudio session </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=05b_count_matrix.html class=md-nav__link> Count matrix </a> </li> <li class=md-nav__item> <a href=05c_count_normalization.html class=md-nav__link> Normalization </a> </li> <li class=md-nav__item> <a href=06_exploratory_analysis.html class=md-nav__link> Exploratory Analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_5> <label class=md-nav__link for=__nav_5_5 id=__nav_5_5_label tabindex=0> Differential Expression <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> Design formula </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> DESeq2 </a> </li> <li class=md-nav__item> <a href=07b_hypothesis_testing.html class=md-nav__link> Hypothesis testing </a> </li> <li class=md-nav__item> <a href=07c_DEA_visualization.html class=md-nav__link> Visualization </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_6> <label class=md-nav__link for=__nav_5_6 id=__nav_5_6_label tabindex=0> Functional Analysis <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_6_label aria-expanded=false> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=08a_FA_genomic_annotation.html class=md-nav__link> Gene annotation </a> </li> <li class=md-nav__item> <a href=08b_FA_overrepresentation.html class=md-nav__link> Over representation analysis </a> </li> <li class=md-nav__item> <a href=08c_FA_GSEA.html class=md-nav__link> Class scoring </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=09_summarized_workflow.html class=md-nav__link> Summarized workflow </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#submit-the-job-in-ucloud class=md-nav__link> Submit the job in Ucloud </a> </li> <li class=md-nav__item> <a href=#stopping-the-app class=md-nav__link> Stopping the app </a> </li> <li class=md-nav__item> <a href=#saved-work class=md-nav__link> Saved work </a> </li> <li class=md-nav__item> <a href=#restarting-the-rstudio-session class=md-nav__link> Restarting the Rstudio session </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=setup-for-teaching-in-ucloud>Setup for teaching in uCloud<a class=headerlink href=#setup-for-teaching-in-ucloud title="Permanent link">&para;</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>&#128368; <strong>Time Estimation:</strong> X minutes </p> <p>&#128172; <strong>Learning Objectives:</strong> </p> <ol> <li>Start a transcriptomics app job in Ucloud for the next lessons in data analysis</li> </ol> </div> <h2 id=submit-the-job-in-ucloud>Submit the job in Ucloud<a class=headerlink href=#submit-the-job-in-ucloud title="Permanent link">&para;</a></h2> <p>Access <a href=https://cloud.sdu.dk>Ucloud</a> with your account and choose the project <code>Sandbox RNASeq Workshop</code> where you have been invited.</p> <p><img alt src=img/05a_data_analysis_setup/chooseProject.png></p> <p>Click on <code>Apps</code> on the left-side menu, and search for the application <code>Transcriptomics Sandbox</code> and click on it.</p> <p><img alt src=img/05a_data_analysis_setup/chooseTranscriptomics.png></p> <p>You will be met with a series of possible parameters to choose. However, we have prepared the parameters already for you! Just click on <code>Import parameters</code>:</p> <p><img alt src=img/05a_data_analysis_setup/importParameters.png></p> <p>Then, <code>Import file from UCloud</code>:</p> <p><img alt src=img/05a_data_analysis_setup/importParameters2.png></p> <p>And select the <code>jobParameters.json</code> in:</p> <ul> <li><code>sandbox_bulkRNASeq</code> -> <code>bulk_RNAseq_course</code> -> <code>Scripts</code> -> <code>ucloud_analysis_setup</code> -> <code>jobParameters.json</code></li> </ul> <div class="admonition warning"> <p class=admonition-title>Warning</p> <p><strong>Make sure that the hard-drive icon says <code>sandbox_bulkRNASeq</code>!!</strong> <img alt src=img/04c_preprocessing_setup/importParameters3.png></p> </div> <p>You are ready to run the app by clicking on the button on the right column of the screen (<code>submit</code>).</p> <p><img alt src=img/05a_data_analysis_setup/submit.png></p> <p>Now, wait some time until the screen looks like the figure below. It usually takes a few minutes for everything to be ready and installed. You can always come back to this screen from the left menu Runs on uCloud, so that you can add extra time or stop the app if you will not use it.</p> <p><img alt src=img/05a_data_analysis_setup/startapp.png></p> <p>Now, click on <code>open interface</code> on the top right-hand side of the screen. You will start Rstudio through your browser!</p> <p>On the lower right side of Rstudio, where you see the file explorer, there shoud be a folder <code>bulk_RNAseq_course</code>. Here you will find the materials of the course, but it is a "read only" file. In order to have a copy for your own purposes, we will use a script. Go to the <strong>Terminal</strong> tab on the top left of the Rstudio session and copy-paste this command:</p> <p><code>./bulk_RNAseq_course/Scripts/ucloud_analysis_setup/ucloud_setup.sh</code></p> <p><img alt src=img/05a_data_analysis_setup/copyMaterial.png></p> <p>Now there should be another folder in the file explorer called <code>introduction_bulkRNAseq_analysis</code>. This is the folder you should use from now on.</p> <p><img alt src=img/05a_data_analysis_setup/courseMaterial.png></p> <p>Now you can go back to the <strong>Console</strong> tab. You are ready to start analysing your data!</p> <h2 id=stopping-the-app>Stopping the app<a class=headerlink href=#stopping-the-app title="Permanent link">&para;</a></h2> <p>When you are done, go on <code>Runs</code> in uCloud, and choose your app if it is still running. Then you will be able to stop it from using resources.</p> <p><img alt src=img/05a_data_analysis_setup/stopRun.png></p> <h2 id=saved-work>Saved work<a class=headerlink href=#saved-work title="Permanent link">&para;</a></h2> <p>After running a first work session, everything that you have created, including the scripts and results of your analysis, will be saved in your own personal <em>"Jobs"</em> folder. Inside this folder there will be a subfolder called <em>Transcriptomics Sandbox</em>, which will contain all the jobs you have run with the Transcriptomics Sandbox app. Inside this folder, you will find your folder named after the job name you gave in the previous step.</p> <ol> <li>Your material will be saved in a volume with your username, that you should be able to see under the menu <code>Files</code>. </li> </ol> <p><img alt src=img/05a_data_analysis_setup/savedWork1.png></p> <ol> <li>Go to <code>Jobs → Transcriptomics Sandbox → job_name → introduction_bulkRNAseq_analysis</code> </li> </ol> <p><img alt src=img/05a_data_analysis_setup/savedWork2.png></p> <h2 id=restarting-the-rstudio-session>Restarting the Rstudio session<a class=headerlink href=#restarting-the-rstudio-session title="Permanent link">&para;</a></h2> <p>If you want to keep working on your previous results, you can restart an Rstudio session following these steps:</p> <p>Click on <code>Apps</code> on the left-side menu, and look for the application <code>Transcriptomics Sandbox</code> and click on it.</p> <p><img alt src=img/05a_data_analysis_setup/chooseTranscriptomics.png></p> <p>You will be met again with a series of possible parameters to choose. You have to assign again the <code>Import parameters</code> file as before.</p> <ul> <li><code>sandbox_bulkRNASeq</code> -> <code>bulk_RNAseq_course</code> -> <code>Scripts</code> -> <code>ucloud_analysis_setup</code> -> <code>jobParameters.json</code></li> </ul> <p><img alt src=img/05a_data_analysis_setup/importParameters.png></p> <p>In <em>"Select folders to use"</em>, add the folder with the results of your previous job:</p> <ul> <li>Go to <code>Member Files: your_username → Jobs → Transcriptomics Sandbox → job_name → introduction_bulkRNAseq_analysis</code>, click "Use."</li> </ul> <p><img alt src=img/05a_data_analysis_setup/restartJob.png></p> <p>You are ready to run the app by clicking on the button on the right column of the screen (<code>submit</code>). After opening the Rstudio interface, you should be able to access the folder <code>introduction_bulkRNAseq_analysis</code>, where you will find your course notebooks and results from your previous work!</p> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=04c_preprocessing_setup.html class="md-footer__link md-footer__link--prev" aria-label="Previous: UCloud setup" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> UCloud setup </div> </div> </a> <a href=05b_count_matrix.html class="md-footer__link md-footer__link--next" aria-label="Next: Count matrix" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Count matrix </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.footer", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.db81ec45.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=assets/javascripts/bundle.a00a7c5e.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>