<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="This course is an introduction for how to approach bulk RNAseq data, starting from the sequencing reads.  It will provide an overview of the fundamentals of RNAseq analysis, including read preprocessing, data normalization, data exploration with PCAs and heatmaps, performing differential expression analysis and annotation of the differentially expressed genes.  Participants will also learn how to evaluate confounding and batch effects in the data.  The course will further touch upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, especially about how they influence downstream bioinformatic analysis.   "><meta name=author content="Jose Alejandro Romero Herrera"><link rel=canonical href=https://hds-sandbox.github.io/bulk_RNAseq_course/08a_FA_genomic_annotation.html><link rel=icon href=img/favicon.png><meta name=generator content="mkdocs-1.4.2, mkdocs-material-8.5.10"><title>Genomic annotations for functional analyses - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.975780f9.min.css><link rel=stylesheet href=assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#genomic-annotations class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Genomic annotations for functional analyses </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class=md-tabs__link> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class=md-tabs__item> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class=md-tabs__item> <a href=05a_data_analysis_setup.html class="md-tabs__link md-tabs__link--active"> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=09_summarized_workflow.html class=md-tabs__link> Summarized workflow </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> Course Introduction </a> </li> <li class=md-nav__item> <a href=02_experimental_planning.html class=md-nav__link> Experimental planning </a> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> Data Explanation </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_4 type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4> Read Preprocessing <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Read Preprocessing" data-md-level=1> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> Preprocessing steps </a> </li> <li class=md-nav__item> <a href=04b_pipelines.html class=md-nav__link> Nextflow and nf-core pipelines </a> </li> <li class=md-nav__item> <a href=04c_preprocessing_setup.html class=md-nav__link> UCloud setup </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5 type=checkbox id=__nav_5 checked> <label class=md-nav__link for=__nav_5> Analysing RNAseq data <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Analysing RNAseq data" data-md-level=1> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=05a_data_analysis_setup.html class=md-nav__link> Ucloud setup </a> </li> <li class=md-nav__item> <a href=05b_count_matrix.html class=md-nav__link> Count matrix </a> </li> <li class=md-nav__item> <a href=05c_count_normalization.html class=md-nav__link> Normalization </a> </li> <li class=md-nav__item> <a href=06_exploratory_analysis.html class=md-nav__link> Exploratory Analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5_5 type=checkbox id=__nav_5_5> <label class=md-nav__link for=__nav_5_5> Differential Expression <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Differential Expression" data-md-level=2> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> Design formula </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> DESeq2 </a> </li> <li class=md-nav__item> <a href=07b_hypothesis_testing.html class=md-nav__link> Hypothesis testing </a> </li> <li class=md-nav__item> <a href=07c_DEA_visualization.html class=md-nav__link> Visualization </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5_6 type=checkbox id=__nav_5_6 checked> <label class=md-nav__link for=__nav_5_6> Functional Analysis <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Functional Analysis" data-md-level=2> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Gene annotation <span class="md-nav__icon md-icon"></span> </label> <a href=08a_FA_genomic_annotation.html class="md-nav__link md-nav__link--active"> Gene annotation </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#databases class=md-nav__link> Databases </a> </li> <li class=md-nav__item> <a href=#genome-builds class=md-nav__link> Genome builds </a> </li> <li class=md-nav__item> <a href=#tools-for-accessing-databases class=md-nav__link> Tools for accessing databases </a> <nav class=md-nav aria-label="Tools for accessing databases"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#annotation-tools-for-accessingquerying-annotations-from-a-specific-databases class=md-nav__link> Annotation tools: for accessing/querying annotations from a specific databases </a> </li> <li class=md-nav__item> <a href=#interface-tools class=md-nav__link> Interface tools </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#annotationdbi class=md-nav__link> AnnotationDbi </a> </li> <li class=md-nav__item> <a href=#annotationhub class=md-nav__link> AnnotationHub </a> <nav class=md-nav aria-label=AnnotationHub> <ul class=md-nav__list> <li class=md-nav__item> <a href=#using-annotationhub-to-create-our-tx2gene-file class=md-nav__link> Using AnnotationHub to create our tx2gene file </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#annotables-package class=md-nav__link> Annotables package </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=08b_FA_overrepresentation.html class=md-nav__link> Over representation analysis </a> </li> <li class=md-nav__item> <a href=08c_FA_GSEA.html class=md-nav__link> Class scoring </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=09_summarized_workflow.html class=md-nav__link> Summarized workflow </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#databases class=md-nav__link> Databases </a> </li> <li class=md-nav__item> <a href=#genome-builds class=md-nav__link> Genome builds </a> </li> <li class=md-nav__item> <a href=#tools-for-accessing-databases class=md-nav__link> Tools for accessing databases </a> <nav class=md-nav aria-label="Tools for accessing databases"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#annotation-tools-for-accessingquerying-annotations-from-a-specific-databases class=md-nav__link> Annotation tools: for accessing/querying annotations from a specific databases </a> </li> <li class=md-nav__item> <a href=#interface-tools class=md-nav__link> Interface tools </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#annotationdbi class=md-nav__link> AnnotationDbi </a> </li> <li class=md-nav__item> <a href=#annotationhub class=md-nav__link> AnnotationHub </a> <nav class=md-nav aria-label=AnnotationHub> <ul class=md-nav__list> <li class=md-nav__item> <a href=#using-annotationhub-to-create-our-tx2gene-file class=md-nav__link> Using AnnotationHub to create our tx2gene file </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#annotables-package class=md-nav__link> Annotables package </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course/edit/master/docs/08a_FA_genomic_annotation.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1 id=genomic-annotations>Genomic annotations<a class=headerlink href=#genomic-annotations title="Permanent link">&para;</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>&#128368; <strong>Time Estimation:</strong> 30 minutes </p> <p>&#128172; <strong>Learning Objectives:</strong> </p> <ol> <li>Discuss the available genomic annotation databases and the different types of information stored</li> <li>Compare and contrast the tools available for accessing genomic annotation databases</li> <li>Apply various R packages for retrieval of genomic annotations</li> </ol> </div> <p>The analysis of next-generation sequencing results requires associating genes, transcripts, proteins, etc. with functional or regulatory information. To perform functional analysis on gene lists, we often need to obtain gene identifiers that are compatible with the tools we wish to use and this is not always trivial. Here, we discuss <strong>ways in which you can obtain gene annotation information and some of the advantages and disadvantages of each method</strong>.</p> <h2 id=databases>Databases<a class=headerlink href=#databases title="Permanent link">&para;</a></h2> <p>We retrieve information on the processes, pathways, etc. (for which a gene is involved in) from the necessary database where the information is stored. The database you choose will be dependent on what type of information you are trying to obtain. Examples of databases that are often queried, include:</p> <p><strong>General databases</strong></p> <p>Offer comprehensive information on genome features, feature coordinates, homology, variant information, phenotypes, protein domain/family information, associated biological processes/pathways, associated microRNAs, etc.:</p> <ul> <li><strong>Ensembl</strong> (use Ensembl gene IDs)</li> <li><strong>NCBI</strong> (use Entrez gene IDs)</li> <li><strong>UCSC</strong></li> <li><strong>EMBL-EBI</strong></li> </ul> <p><strong>Annotation-specific databases</strong></p> <p>Provide annotations related to a specific topic:</p> <ul> <li><strong>Gene Ontology (GO):</strong> database of gene ontology biological processes, cellular components and molecular functions - based on Ensembl or Entrez gene IDs or official gene symbols</li> <li><strong>KEGG:</strong> database of biological pathways - based on Entrez gene IDs</li> <li><strong>MSigDB:</strong> database of gene sets</li> <li><strong>Reactome:</strong> database of biological pathways</li> <li><strong>Human Phenotype Ontology:</strong> database of genes associated with human disease</li> <li><strong>CORUM:</strong> database of protein complexes for human, mouse, rat</li> <li><strong>…</strong></li> </ul> <p>This is by no means an exhaustive list, there are many other databases available that are not listed here.</p> <h2 id=genome-builds>Genome builds<a class=headerlink href=#genome-builds title="Permanent link">&para;</a></h2> <p>Before you begin your search through any of these databases, you should know which <strong>build of the genome</strong> was used to generate your gene list and make sure you use the <strong>same build for the annotations</strong> during functional analysis. When a new genome build is acquired, the names and/or coordinate location of genomic features (gene, transcript, exon, etc.) may change. Therefore, the annotations regarding genome features (gene, transcript, exon, etc.) is genome-build specific and we need to make sure that our annotations are obtained from the appropriate resource.</p> <p>For example, if we used the GRCh38 build of the human genome to quantify gene expression used for differential expression analysis, then we should use the <strong>same GRCh38 build</strong> of the genome to convert between gene IDs and to identify annotations for each of the genes.</p> <h2 id=tools-for-accessing-databases>Tools for accessing databases<a class=headerlink href=#tools-for-accessing-databases title="Permanent link">&para;</a></h2> <p>Within R, there are many popular packages used for gene/transcript-level annotation. These packages provide tools that take the list of genes you provide and retrieve information for each gene using one or more of the databases listed above.</p> <h3 id=annotation-tools-for-accessingquerying-annotations-from-a-specific-databases>Annotation tools: for accessing/querying annotations from a specific databases<a class=headerlink href=#annotation-tools-for-accessingquerying-annotations-from-a-specific-databases title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th align=center>Tool</th> <th align=left>Description</th> <th align=left>Pros</th> <th align=center>Cons</th> </tr> </thead> <tbody> <tr> <td align=center><strong><a href=https://bioconductor.org/packages/release/bioc/vignettes/AnnotationDbi/inst/doc/IntroToAnnotationPackages.pdf>org.Xx.eg.db</a></strong></td> <td align=left>Query gene feature information for the organism of interest</td> <td align=left>gene ID conversion, biotype and coordinate information</td> <td align=center>only latest genome build available</td> </tr> <tr> <td align=center><strong><a href=http://bioconductor.org/packages/devel/bioc/vignettes/ensembldb/inst/doc/ensembldb.html>EnsDb.Xx.vxx</a></strong></td> <td align=left>Transcript and gene-level information directly fetched from Ensembl API (similar to TxDb, but with filtering ability and versioned by Ensembl release)</td> <td align=left>easy functions to extract features, direct filtering</td> <td align=center>Not the most up-to-date annotations, more difficult to use than some packages</td> </tr> <tr> <td align=center><strong><a href=https://bioconductor.org/packages/release/bioc/vignettes/GenomicFeatures/inst/doc/GenomicFeatures.pdf>TxDb.Xx.UCSC.hgxx.knownGene</a></strong></td> <td align=left>UCSC database for transcript and gene-level information or can create own <em>TxDb</em> from an SQLite database file using the <em>GenomicFeatures</em> package</td> <td align=left>feature information, easy functions to extract features</td> <td align=center>only available current and recent genome builds - can create your own, less up-to-date with annotations than Ensembl</td> </tr> <tr> <td align=center><strong><a href=https://github.com/stephenturner/annotables>annotables</a></strong></td> <td align=left>Gene-level feature information immediately available for the human and model organisms</td> <td align=left>super quick and easy gene ID conversion, biotype and coordinate information</td> <td align=center>static resource, not updated regularly</td> </tr> <tr> <td align=center><strong><a href=https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/biomaRt.html>biomaRt</a></strong></td> <td align=left>An R package version of the Ensembl <a href=http://www.ensembl.org/biomart/martview/70dbbbe3f1c5389418b5ea1e02d89af3>BioMart online tool</a></td> <td align=left>all Ensembl database information available, all organisms on Ensembl, wealth of information</td> <td align=center></td> </tr> </tbody> </table> <h3 id=interface-tools>Interface tools<a class=headerlink href=#interface-tools title="Permanent link">&para;</a></h3> <p>Other packages are design for accessing/querying annotations from multiple different annotation sources</p> <ul> <li><strong>AnnotationDbi:</strong> queries the <em>OrgDb</em>, <em>TxDb</em>, <em>Go.db</em>, <em>EnsDb</em>, and <em>BioMart</em> annotations. </li> <li><strong>AnnotationHub:</strong> queries large collection of whole genome resources, including ENSEMBL, UCSC, ENCODE, Broad Institute, KEGG, NIH Pathway Interaction Database, etc.</li> </ul> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>These are both packages that can be used to create the <code>tx2gene</code> files that salmon gave us in case you did not have them.</p> </div> <h2 id=annotationdbi>AnnotationDbi<a class=headerlink href=#annotationdbi title="Permanent link">&para;</a></h2> <p>AnnotationDbi is an R package that provides an interface for connecting and querying various annotation databases using SQLite data storage. The AnnotationDbi packages can query the <em>OrgDb</em>, <em>TxDb</em>, <em>EnsDb</em>, <em>Go.db</em>, and <em>BioMart</em> annotations. There is helpful <a href=https://bioconductor.org/packages/release/bioc/vignettes/AnnotationDbi/inst/doc/IntroToAnnotationPackages.pdf>documentation</a> available to reference when extracting data from any of these databases.</p> <h2 id=annotationhub>AnnotationHub<a class=headerlink href=#annotationhub title="Permanent link">&para;</a></h2> <p>AnnotationHub is a wonderful resource for accessing genomic data or querying large collection of whole genome resources, including ENSEMBL, UCSC, ENCODE, Broad Institute, KEGG, NIH Pathway Interaction Database, etc. All of this information is stored and easily accessible by directly connecting to the database.</p> <p>To get started with AnnotationHub, we first load the library and connect to the database:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1># Load libraries</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=nf>library</span><span class=p>(</span><span class=n>AnnotationHub</span><span class=p>)</span>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=nf>library</span><span class=p>(</span><span class=n>ensembldb</span><span class=p>)</span>
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a>
<a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a><span class=c1># Connect to AnnotationHub</span>
<a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a><span class=n>ah</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>AnnotationHub</span><span class=p>()</span>
</code></pre></div> <div class="admonition warning"> <p class=admonition-title>Warning</p> <p>The script will ask you to create a cache directory, type yes!</p> </div> <div class="admonition info"> <p class=admonition-title>What is a cache?</p> <p>A cache is used in R to store data or a copy of the data so that future requests can be served faster without having to re-run a lengthy computation.</p> <p>The <code>AnnotationHub()</code> command creates a client that manages a local cache of the database, helping with quick and reproducible access. When encountering question <code>AnnotationHub does not exist, create directory?</code>, you can anwser either <code>yes</code> (create a permanent location to store cache) or <code>no</code> (create a temporary location to store cache). <code>hubCache(ah)</code> gets the file system location of the local AnnotationHub cache. <code>hubUrl(ah)</code> gets the URL for the online hub. </p> </div> <p>To see the types of information stored inside our database, we can just type the name of the object:</p> <div class="admonition note"> <p class=admonition-title>Note</p> <p>Results here will differ from yours</p> </div> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1># Explore the AnnotationHub object</span>
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a><span class=n>ah</span>
</code></pre></div> <p>Using the output, you can get an idea of the information that you can query within the AnnotationHub object.</p> <div class=highlight><pre><span></span><code>AnnotationHub with 47240 records
# snapshotDate(): 2019-10-29 
# $dataprovider: BroadInstitute, Ensembl, UCSC, ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/, H...
# $species: Homo sapiens, Mus musculus, Drosophila melanogaster, Bos taurus, Pan troglod...
# $rdataclass: GRanges, BigWigFile, TwoBitFile, Rle, OrgDb, EnsDb, ChainFile, TxDb, Inpa...
# additional mcols(): taxonomyid, genome, description, coordinate_1_based,
#   maintainer, rdatadateadded, preparerclass, tags, rdatapath, sourceurl,
#   sourcetype 
# retrieve records with, e.g., &#39;object[[&quot;AH5012&quot;]]&#39;

            title                                                                   
  AH5012  | Chromosome Band                                                         
  AH5013  | STS Markers                                                             
  AH5014  | FISH Clones                                                             
  AH5015  | Recomb Rate                                                             
  AH5016  | ENCODE Pilot                                                            
  ...       ...                                                                     
  AH78364 | Xiphophorus_maculatus.X_maculatus-5.0-male.ncrna.2bit                   
  AH78365 | Zonotrichia_albicollis.Zonotrichia_albicollis-1.0.1.cdna.all.2bit       
  AH78366 | Zonotrichia_albicollis.Zonotrichia_albicollis-1.0.1.dna_rm.toplevel.2bit
  AH78367 | Zonotrichia_albicollis.Zonotrichia_albicollis-1.0.1.dna_sm.toplevel.2bit
  AH78368 | Zonotrichia_albicollis.Zonotrichia_albicollis-1.0.1.ncrna.2bit
</code></pre></div> <p>Notice the note on retrieving records with <code>object[["AH5012"]]</code> - this will be how we can <strong>extract a single record</strong> from the AnnotationHub object.</p> <p>If you would like to see more information about any of the classes of data you can extract that information as well. For example, if you wanted to <strong>determine all species information available</strong>, you could explore that within the AnnotationHub object:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=c1># Explore all species information available</span>
<a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=nf>unique</span><span class=p>(</span><span class=n>ah</span><span class=o>$</span><span class=n>species</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>head</span><span class=p>()</span>
</code></pre></div> <p>In addition to species information, there is also additional information about the type of Data Objects and the Data Providers:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=c1># Explore the types of Data Objects available</span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=nf>unique</span><span class=p>(</span><span class=n>ah</span><span class=o>$</span><span class=n>rdataclass</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>head</span><span class=p>()</span>
<a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a>
<a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a><span class=c1># Explore the Data Providers</span>
<a id=__codelineno-3-5 name=__codelineno-3-5 href=#__codelineno-3-5></a><span class=nf>unique</span><span class=p>(</span><span class=n>ah</span><span class=o>$</span><span class=n>dataprovider</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>head</span><span class=p>()</span>
</code></pre></div> <p>Now that we know the types of information available from AnnotationHub we can query it for the information we want using the <code>query()</code> function. Let’s say we would like to <strong>return the Ensembl <code>EnsDb</code> information for Human</strong>. To return the records available, we need to use the terms as they are output from the <code>ah</code> object to extract the desired data.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=c1># Query AnnotationHub</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=n>human_ens</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>query</span><span class=p>(</span><span class=n>ah</span><span class=p>,</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;Homo sapiens&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;EnsDb&quot;</span><span class=p>))</span>
</code></pre></div> <p>The query retrieves all <strong>hits for the <code>EnsDb</code> objects</strong>, and you will see that they are listed by the release number. The most current release for GRCh38 is Ensembl98 and AnnotationHub offers that as an option to use. However, if you look at options for older releases, for Homo sapiens it only go back as far as Ensembl 87. This is fine if you are using GRCh38, however if you were using an older genome build like hg19/GRCh37, you would need to load the <code>EnsDb</code> package if available for that release or you might need to build your own with <code>ensembldb</code>.</p> <div class="admonition note"> <p class=admonition-title>Note</p> <p>Results here will differ from yours</p> </div> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=n>human_ens</span>
</code></pre></div> <div class=highlight><pre><span></span><code>AnnotationHub with 13 records
# snapshotDate(): 2019-10-29 
# $dataprovider: Ensembl
# $species: Homo sapiens
# $rdataclass: EnsDb
# additional mcols(): taxonomyid, genome, description, coordinate_1_based,
#   maintainer, rdatadateadded, preparerclass, tags, rdatapath, sourceurl,
#   sourcetype 
# retrieve records with, e.g., &#39;object[[&quot;AH53211&quot;]]&#39;

            title                            
  AH53211 | Ensembl 87 EnsDb for Homo Sapiens
  AH53715 | Ensembl 88 EnsDb for Homo Sapiens
  AH56681 | Ensembl 89 EnsDb for Homo Sapiens
  AH57757 | Ensembl 90 EnsDb for Homo Sapiens
  AH60773 | Ensembl 91 EnsDb for Homo Sapiens
  ...       ...                              
  AH67950 | Ensembl 95 EnsDb for Homo sapiens
  AH69187 | Ensembl 96 EnsDb for Homo sapiens
  AH73881 | Ensembl 97 EnsDb for Homo sapiens
  AH73986 | Ensembl 79 EnsDb for Homo sapiens
  AH75011 | Ensembl 98 EnsDb for Homo sapiens
</code></pre></div> <p>In our case, we are looking for the latest Ensembl release so that the annotations are the most up-to-date. To extract this information from AnnotationHub, we can use the AnnotationHub ID to <strong>subset the object</strong>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=c1># Extract annotations of interest</span>
<a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=n>human_ens</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>human_ens</span><span class=p>[[</span><span class=nf>length</span><span class=p>(</span><span class=n>human_ens</span><span class=p>)]]</span><span class=w> </span><span class=c1># We extract latest</span>
</code></pre></div> <p>Now we can use <code>ensembldb</code> functions to extract the information at the gene, transcript, or exon levels. We are interested in the gene-level annotations, so we can extract that information as follows:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1># Extract gene-level information</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=nf>genes</span><span class=p>(</span><span class=n>human_ens</span><span class=p>,</span><span class=w> </span><span class=n>return.type</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;data.frame&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>head</span><span class=p>()</span>
</code></pre></div> <p>But note that it is just as easy to get the transcript- or exon-level information:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=c1># Extract transcript-level information</span>
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=nf>transcripts</span><span class=p>(</span><span class=n>human_ens</span><span class=p>,</span><span class=w> </span><span class=n>return.type</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;data.frame&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>head</span><span class=p>()</span>
<a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a>
<a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=c1># Extract exon-level information</span>
<a id=__codelineno-8-5 name=__codelineno-8-5 href=#__codelineno-8-5></a><span class=nf>exons</span><span class=p>(</span><span class=n>human_ens</span><span class=p>,</span><span class=w> </span><span class=n>return.type</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;data.frame&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>head</span><span class=p>()</span>
</code></pre></div> <p>To <strong>obtain an annotation data frame</strong> using AnnotationHub, we’ll use the <code>genes()</code> function, but only keep selected columns and filter out rows to keep those corresponding to our gene identifiers in our results file:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=c1># Create a gene-level dataframe </span>
<a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a><span class=n>annotations_ahb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>genes</span><span class=p>(</span><span class=n>human_ens</span><span class=p>,</span><span class=w> </span><span class=n>return.type</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;data.frame&quot;</span><span class=p>)</span><span class=w>  </span><span class=o>%&gt;%</span>
<a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a><span class=w>  </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>gene_id</span><span class=p>,</span><span class=w> </span><span class=n>gene_name</span><span class=p>,</span><span class=w> </span><span class=n>entrezid</span><span class=p>,</span><span class=w> </span><span class=n>gene_biotype</span><span class=p>,</span><span class=w> </span><span class=n>description</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-9-4 name=__codelineno-9-4 href=#__codelineno-9-4></a><span class=w>  </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>gene_id</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=n>res_tableOE_tb</span><span class=o>$</span><span class=n>gene</span><span class=p>)</span>
</code></pre></div> <p>This dataframe looks like it should be fine as it is, but we look a little closer we will notice that the column containing Entrez identifiers is a list, and in fact there are many Ensembl identifiers that map to more than one Entrez identifier!</p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=c1># Wait a second, we don&#39;t have one-to-one mappings!</span>
<a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=nf>class</span><span class=p>(</span><span class=n>annotations_ahb</span><span class=o>$</span><span class=n>entrezid</span><span class=p>)</span>
<a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a><span class=nf>which</span><span class=p>(</span><span class=nf>map</span><span class=p>(</span><span class=n>annotations_ahb</span><span class=o>$</span><span class=n>entrezid</span><span class=p>,</span><span class=w> </span><span class=n>length</span><span class=p>)</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=m>1</span><span class=p>)</span>
</code></pre></div> <p>So what do we do here? And why do we have this problem? An answer from the <a href=https://www.biostars.org/p/16505/ >Ensembl Help Desk</a> is that this occurs when we cannot choose a perfect match; ie when we have two good matches, but one does not appear to match with a better percentage than the other. In that case, we assign both matches. What we will do is choose to <strong>keep the first identifier for these multiple mapping cases</strong>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=n>annotations_ahb</span><span class=o>$</span><span class=n>entrezid</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>map</span><span class=p>(</span><span class=n>annotations_ahb</span><span class=o>$</span><span class=n>entrezid</span><span class=p>,</span><span class=m>1</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w>  </span><span class=nf>unlist</span><span class=p>()</span>
</code></pre></div> <div class="admonition info"> <p class=admonition-title>Info</p> <p>Not all databases handle multiple mappings in the same way. For example, if we used the OrgDb instead of the EnsDb:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=n>human_orgdb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>query</span><span class=p>(</span><span class=n>ah</span><span class=p>,</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;Homo sapiens&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;OrgDb&quot;</span><span class=p>))</span>
<a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a><span class=n>human_orgdb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>human_ens</span><span class=p>[[</span><span class=nf>length</span><span class=p>(</span><span class=n>human_ens</span><span class=p>)]]</span>
<a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a><span class=n>annotations_orgdb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>select</span><span class=p>(</span><span class=n>human_orgdb</span><span class=p>,</span><span class=w> </span><span class=n>res_tableOE_tb</span><span class=o>$</span><span class=n>gene</span><span class=p>,</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;SYMBOL&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;GENENAME&quot;</span><span class=p>,</span><span class=w>         </span><span class=s>&quot;ENTREZID&quot;</span><span class=p>),</span><span class=w> </span><span class=s>&quot;ENSEMBL&quot;</span><span class=p>)</span>
</code></pre></div> <p>We would find that multiple mapping entries would be automatically reduced to one-to-one. We would also find that more than half of the input genes do not return any annotations. This is because the OrgDb family of database are primarily based on mapping using Entrez Gene identifiers. Since our data is based on Ensembl mappings, using the OrgDb would result in a loss of information.</p> </div> <p>Let’s take a look and see how many of our Ensembl identifiers have an associated gene symbol, and how many of them are unique:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=nf>which</span><span class=p>(</span><span class=nf>is.na</span><span class=p>(</span><span class=n>annotations_ahb</span><span class=o>$</span><span class=n>gene_name</span><span class=p>))</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>length</span><span class=p>()</span>
<a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a>
<a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a><span class=nf>which</span><span class=p>(</span><span class=nf>duplicated</span><span class=p>(</span><span class=n>annotations_ahb</span><span class=o>$</span><span class=n>gene_name</span><span class=p>))</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>length</span><span class=p>()</span>
</code></pre></div> <p>Let’s identify the non-duplicated genes and only keep the ones that are not duplicated:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=c1># Determine the indices for the non-duplicated genes</span>
<a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a><span class=n>non_duplicates_idx</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>which</span><span class=p>(</span><span class=nf>duplicated</span><span class=p>(</span><span class=n>annotations_ahb</span><span class=o>$</span><span class=n>gene_name</span><span class=p>)</span><span class=w> </span><span class=o>==</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a>
<a id=__codelineno-14-4 name=__codelineno-14-4 href=#__codelineno-14-4></a><span class=c1># How many rows does annotations_ahb have?</span>
<a id=__codelineno-14-5 name=__codelineno-14-5 href=#__codelineno-14-5></a><span class=n>annotations_ahb</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>nrow</span><span class=p>()</span>
<a id=__codelineno-14-6 name=__codelineno-14-6 href=#__codelineno-14-6></a>
<a id=__codelineno-14-7 name=__codelineno-14-7 href=#__codelineno-14-7></a><span class=c1># Return only the non-duplicated genes using indices</span>
<a id=__codelineno-14-8 name=__codelineno-14-8 href=#__codelineno-14-8></a><span class=n>annotations_ahb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>annotations_ahb</span><span class=p>[</span><span class=n>non_duplicates_idx</span><span class=p>,</span><span class=w> </span><span class=p>]</span>
<a id=__codelineno-14-9 name=__codelineno-14-9 href=#__codelineno-14-9></a>
<a id=__codelineno-14-10 name=__codelineno-14-10 href=#__codelineno-14-10></a><span class=c1># How many rows are we left with after removing?</span>
<a id=__codelineno-14-11 name=__codelineno-14-11 href=#__codelineno-14-11></a><span class=n>annotations_ahb</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>nrow</span><span class=p>()</span>
</code></pre></div> <p>Finally, it would be good to know <strong>what proportion of the Ensembl identifiers map to an Entrez identifier</strong>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=c1># Determine how many of the Entrez column entries are NA</span>
<a id=__codelineno-15-2 name=__codelineno-15-2 href=#__codelineno-15-2></a><span class=nf>which</span><span class=p>(</span><span class=nf>is.na</span><span class=p>(</span><span class=n>annotations_ahb</span><span class=o>$</span><span class=n>entrezid</span><span class=p>))</span><span class=w> </span><span class=o>%&gt;%</span><span class=w>  </span><span class=nf>length</span><span class=p>()</span>
</code></pre></div> <p>That’s more than half of our genes! If we plan on using Entrez ID results for downstream analysis, we should definitely keep this in mind. If you look at some of the Ensembl IDs from our query that returned NA, these map to pseudogenes (i.e <a href="https://useast.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000265439;r=6:44209766-44210063;t=ENST00000580735">ENSG00000265439</a>) or non-coding RNAs (i.e. <a href="http://useast.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000265425;r=18:68427030-68436918;t=ENST00000577835">ENSG00000265425</a>). The discrepancy (which we can expect to observe) between databases is due to the fact that each implements its own different computational approaches for generating the gene builds.</p> <h3 id=using-annotationhub-to-create-our-tx2gene-file>Using AnnotationHub to create our tx2gene file<a class=headerlink href=#using-annotationhub-to-create-our-tx2gene-file title="Permanent link">&para;</a></h3> <p>To create our <code>tx2gene</code> file, we would need to use a combination of the methods above and merge two dataframes together. For example:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a><span class=c1>## DO NOT RUN THIS CODE</span>
<a id=__codelineno-16-2 name=__codelineno-16-2 href=#__codelineno-16-2></a>
<a id=__codelineno-16-3 name=__codelineno-16-3 href=#__codelineno-16-3></a><span class=c1># Create a transcript dataframe</span>
<a id=__codelineno-16-4 name=__codelineno-16-4 href=#__codelineno-16-4></a><span class=w> </span><span class=n>txdb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>transcripts</span><span class=p>(</span><span class=n>human_ens</span><span class=p>,</span><span class=w> </span><span class=n>return.type</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;data.frame&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-16-5 name=__codelineno-16-5 href=#__codelineno-16-5></a><span class=w>   </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>tx_id</span><span class=p>,</span><span class=w> </span><span class=n>gene_id</span><span class=p>)</span>
<a id=__codelineno-16-6 name=__codelineno-16-6 href=#__codelineno-16-6></a><span class=w> </span><span class=n>txdb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>txdb</span><span class=p>[</span><span class=nf>grep</span><span class=p>(</span><span class=s>&quot;ENST&quot;</span><span class=p>,</span><span class=w> </span><span class=n>txdb</span><span class=o>$</span><span class=n>tx_id</span><span class=p>),]</span>
<a id=__codelineno-16-7 name=__codelineno-16-7 href=#__codelineno-16-7></a>
<a id=__codelineno-16-8 name=__codelineno-16-8 href=#__codelineno-16-8></a><span class=w> </span><span class=c1># Create a gene-level dataframe</span>
<a id=__codelineno-16-9 name=__codelineno-16-9 href=#__codelineno-16-9></a><span class=w> </span><span class=n>genedb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>genes</span><span class=p>(</span><span class=n>human_ens</span><span class=p>,</span><span class=w> </span><span class=n>return.type</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;data.frame&quot;</span><span class=p>)</span><span class=w>  </span><span class=o>%&gt;%</span>
<a id=__codelineno-16-10 name=__codelineno-16-10 href=#__codelineno-16-10></a><span class=w>   </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>gene_id</span><span class=p>,</span><span class=w> </span><span class=n>gene_name</span><span class=p>)</span>
<a id=__codelineno-16-11 name=__codelineno-16-11 href=#__codelineno-16-11></a>
<a id=__codelineno-16-12 name=__codelineno-16-12 href=#__codelineno-16-12></a><span class=w> </span><span class=c1># Merge the two dataframes together</span>
<a id=__codelineno-16-13 name=__codelineno-16-13 href=#__codelineno-16-13></a><span class=w> </span><span class=n>annotations</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>inner_join</span><span class=p>(</span><span class=n>txdb</span><span class=p>,</span><span class=w> </span><span class=n>genedb</span><span class=p>)</span>
</code></pre></div> <p>In this lesson our focus has been using annotation packages to extract information mainly just for gene ID conversion for the different tools that we use downstream. Many of the annotation packages we have presented have much more information than what we need for functional analysis and we have only just scratched the surface here. It’s good to know the capabilities of the tools we use, so we encourage you to spend some time exploring these packages to become more familiar with them.</p> <h2 id=annotables-package>Annotables package<a class=headerlink href=#annotables-package title="Permanent link">&para;</a></h2> <p>The <em>annotables</em> package is a super easy annotation package to use. It is not updated frequently, so it’s not great for getting the most up-to-date information for the current builds and does not have information for other organisms than human and mouse, but is a quick way to get annotation information.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a><span class=c1># Install package</span>
<a id=__codelineno-17-2 name=__codelineno-17-2 href=#__codelineno-17-2></a><span class=n>BiocManager</span><span class=o>::</span><span class=nf>install</span><span class=p>(</span><span class=s>&quot;annotables&quot;</span><span class=p>)</span>
<a id=__codelineno-17-3 name=__codelineno-17-3 href=#__codelineno-17-3></a>
<a id=__codelineno-17-4 name=__codelineno-17-4 href=#__codelineno-17-4></a><span class=c1># Load library</span>
<a id=__codelineno-17-5 name=__codelineno-17-5 href=#__codelineno-17-5></a><span class=nf>library</span><span class=p>(</span><span class=n>annotables</span><span class=p>)</span>
<a id=__codelineno-17-6 name=__codelineno-17-6 href=#__codelineno-17-6></a>
<a id=__codelineno-17-7 name=__codelineno-17-7 href=#__codelineno-17-7></a><span class=c1># Access previous build of annotations</span>
<a id=__codelineno-17-8 name=__codelineno-17-8 href=#__codelineno-17-8></a><span class=n>grch37</span>
</code></pre></div> <p>We can see that the <code>grch37</code> object already contains all the information we want in a super easy way. Let’s annotate the results of our shrunken DEA for MOV10 overexpression:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a><span class=c1>## Re-run this code if you are unsure that you have the right table</span>
<a id=__codelineno-18-2 name=__codelineno-18-2 href=#__codelineno-18-2></a><span class=n>res_tableOE</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>lfcShrink</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>coef</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;sampletype_MOV10_overexpression_vs_control&quot;</span><span class=p>)</span>
<a id=__codelineno-18-3 name=__codelineno-18-3 href=#__codelineno-18-3></a><span class=n>res_tableOE_tb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tableOE</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-18-4 name=__codelineno-18-4 href=#__codelineno-18-4></a><span class=w>    </span><span class=nf>data.frame</span><span class=p>()</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-18-5 name=__codelineno-18-5 href=#__codelineno-18-5></a><span class=w>    </span><span class=nf>rownames_to_column</span><span class=p>(</span><span class=n>var</span><span class=o>=</span><span class=s>&quot;gene&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-18-6 name=__codelineno-18-6 href=#__codelineno-18-6></a><span class=w>    </span><span class=nf>as_tibble</span><span class=p>()</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a><span class=c1>## Return the IDs for the gene symbols in the DE results</span>
<a id=__codelineno-19-2 name=__codelineno-19-2 href=#__codelineno-19-2></a><span class=n>ids</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>grch37</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>ensgene</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=nf>rownames</span><span class=p>(</span><span class=n>res_tableOE</span><span class=p>))</span>
<a id=__codelineno-19-3 name=__codelineno-19-3 href=#__codelineno-19-3></a>
<a id=__codelineno-19-4 name=__codelineno-19-4 href=#__codelineno-19-4></a><span class=c1>## Merge the IDs with the results </span>
<a id=__codelineno-19-5 name=__codelineno-19-5 href=#__codelineno-19-5></a><span class=n>res_ids</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>inner_join</span><span class=p>(</span><span class=n>res_tableOE_tb</span><span class=p>,</span><span class=w> </span><span class=n>ids</span><span class=p>,</span><span class=w> </span><span class=n>by</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;gene&quot;</span><span class=o>=</span><span class=s>&quot;ensgene&quot;</span><span class=p>))</span>
<a id=__codelineno-19-6 name=__codelineno-19-6 href=#__codelineno-19-6></a>
<a id=__codelineno-19-7 name=__codelineno-19-7 href=#__codelineno-19-7></a><span class=nf>head</span><span class=p>(</span><span class=n>res_ids</span><span class=p>)</span>
</code></pre></div> <p>Our data is now ready to use for functional analysis! We have all the ids necessary to proceed.</p> <div class="admonition question"> <p class=admonition-title><strong>Exercise 1</strong></p> <ul> <li>Create a new <code>res_ids</code> object using the <code>annotables</code> package with the human build grch38. <strong>NOTE</strong> call it <code>res_ids_grch38</code>!</li> <li>What are the differences between the <code>res_id_ahb</code>object and the <code>res_ids_grch38</code>?</li> </ul> </div> <div class="admonition question"> <p class=admonition-title><strong>Exercise 2</strong></p> <p>Annotate the results of your DEA for knockdown vs control </p> </div> <hr> <p><em>This lesson was originally developed by members of the teaching team (Mary Piper) at the <a href=http://bioinformatics.sph.harvard.edu/ >Harvard Chan Bioinformatics Core (HBC)</a>.</em></p> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=07c_DEA_visualization.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Visualization" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Visualization </div> </div> </a> <a href=08b_FA_overrepresentation.html class="md-footer__link md-footer__link--next" aria-label="Next: Over representation analysis" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Over representation analysis </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=assets/javascripts/bundle.5a2dcb6a.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>