---
title: "Summary of DGE workflow"
author: "Jose Alejandro Romero Herrera"
date: '`r Sys.Date()`'
knit: (function(inputFile, encoding) { 
      rmarkdown::render(inputFile,
                        encoding=encoding,
                        output_format='all',
                        output_dir='../')})
output:
  # To create PDF report, uncomment below
  #pdf_document:
  #  toc: yes
  md_document:
    variant: gfm
---

```{r knitr, include = FALSE}
DOCNAME = knitr::current_input()
DOCNAME = gsub(DOCNAME, pattern = ".Rmd", replacement = "", fixed = T)
knitr::opts_chunk$set(autodep        = TRUE,
                      cache          = FALSE,
                      cache.path     = paste0("cache/", DOCNAME, "/"),
                      cache.comments = TRUE,
                      echo           = TRUE,
                      error          = FALSE,
                      fig.align      = "center",
                      fig.path       = paste0("./img/", DOCNAME, "/"),
                      #fig.width      = 10,
                      #fig.height     = 8,
                      message        = FALSE,
                      warning        = FALSE,
                      eval           = FALSE)
```

Approximate time: 15 minutes

## Learning Objectives 

* Identify the R commands needed to run a complete differential expression analysis using DESeq2

## Summary of differential expression analysis workflow

We have detailed the various steps in a differential expression analysis workflow, providing theory with example code. To provide a more succinct reference for the code needed to run a DGE analysis, we have summarized the steps in an analysis below:

1. Obtaining gene-level counts from your preprocessing

```{r}
data <- read_table("data/Mov10_full_counts.txt") 

meta <- read_table("meta/Mov10_full_meta.txt")
```

2. Creating the dds object:

```{r}
# Check that the row names of the metadata equal the column names of the **raw counts** data
all(colnames(datacounts)[-1] == meta$samplename)

# Create DESeq2Dataset object
dds <- DESeqDataSetFromMatrix(countData = data.frame(data[,-1], row.names = data$GeneSymbol), 
                              colData = meta, 
                              design = ~ sampletype)
```

3. Exploratory data analysis (PCA & hierarchical clustering) - identifying outliers and sources of variation in the data:

```{r}
# Transform counts for data visualization
rld <- rlog(dds, 
            blind=TRUE)

# Plot PCA 
plotPCA(rld, 
        intgroup="sampletype")

# Extract the rlog matrix from the object and compute pairwise correlation values
rld_mat <- assay(rld)
rld_cor <- cor(rld_mat)

# Plot heatmap
pheatmap(rld_cor, 
         annotation = meta %>% column_to_rownames("samplename"))
```

4. Run DESeq2:

```{r}
# **Optional step** - Re-create DESeq2 dataset if the design formula has changed after QC analysis in include other sources of variation using "dds <- DESeqDataSetFromMatrix(data, colData = metadata, design = ~ covariate + condition)"

# Run DESeq2 differential expression analysis
dds <- DESeq(dds)

# **Optional step** - Output normalized counts to save as a file to access outside RStudio using "normalized_counts <- counts(dds, normalized=TRUE)"
```

5. Check the fit of the dispersion estimates:

```{r}
# Plot dispersion estimates
plotDispEsts(dds)
``` 

6. Create contrasts to perform Wald testing on the shrunken log2 foldchanges between specific conditions:

```{r}
# Specify contrast for comparison of interest
contrast <- c("condition", "level_to_compare", "base_level")

# Output results of Wald test for contrast
res <- results(dds, 
               contrast = contrast, 
               alpha = 0.05)

# Shrink the log2 fold changes to be more accurate
res <- lfcShrink(dds, 
                 contrast = contrast, 
                 type = "normal")
```

7. Output significant results:

```{r}
# Set thresholds
padj.cutoff < - 0.05

# Turn the results object into a tibble for use with tidyverse functions
res_tbl <- res %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

# Subset the significant results
sig_res <- filter(res_tbl, 
                  padj < padj.cutoff)
```

8. Visualize results: volcano plots, heatmaps, normalized counts plots of top genes, etc.

9. Perform analysis to extract functional significance of results: GO or KEGG enrichment, GSEA, etc.

10. Make sure to output the versions of all tools used in the DE analysis:

```{r}
sessionInfo()
```

For better reproducibility, it can help to create **RMarkdown reports**, which save all code, results, and visualizations as nicely formatted html reports. We have available an [example html report](https://www.dropbox.com/s/xp7l0b46rvwmbya/workshop-example.html?dl=1) for perusal. To create these reports we have [additional materials](https://hbctraining.github.io/Training-modules/Rmarkdown/) available.

***

*This lesson was originally developed by members of the teaching team (Mary Piper) at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/).*